---
title: "Putative Freshwater Genes in Pipefish"
author: "Sarah P. Flanagan"
date: "July 25, 2017"
output: 
  html_document:
    fig_caption: true
---


```{r setup}
knitr::opts_knit$set(root.dir='../fwsw_results/')
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
library(scales)

```

## Freshwater putative genes

This analysis supplements that found in **fwsw_analysis.R**, which runs the actual analysis of the dataset. Here, I compare the Fst outliers and delta-divergence outliers to putative genes identified from other studies of teleosts adapting to freshwater environments.

### Read in all the files
```{r files}
#putative genes
put.genes<-read.delim("putative_genes.txt",header=TRUE,sep='\t')
#genome annotations
gff.name<-list.files(path = "../../scovelli_genome/",pattern="ssc_\\d+.*_chromlevel.gff")
if(length(grep("gz",gff.name))>0){
  gff<-read.delim(gzfile(paste("../../scovelli_genome/",gff.name,sep="")),header=F)
} else{
  gff<-read.delim(paste("../../scovelli_genome/",gff.name,sep=""),header=F)
}
colnames(gff)<-c("seqname","source","feature","start","end","score","strand","frame","attribute")
#Stacks output
vcf<-parse.vcf("p4.upd.vcf")
vcf$SNP<-paste(vcf$`#CHROM`,vcf$POS,sep=".")
scaffs<-levels(as.factor(vcf[,1]))
fwsw.tx<-read.delim("stacks/p4.fst_TXCC-TXFW.txt")
fwsw.la<-read.delim("stacks/p4.fst_ALST-LAFW.txt")
fwsw.al<-read.delim("stacks/p4.fst_ALFW-ALST.txt")
fwsw.fl<-read.delim("stacks/p4.fst_FLCC-FLLG.txt")
stacks.sig<-read.delim("p4.stacks.sig.snps.txt")
#annotated Stacks outliers at 0.01
fw.sig.reg<-read.csv("p4.StacksFWSWOutliers_annotatedByGenome.csv")
#deltad
deltad<-read.delim("sep.deltadivergence.txt")
sdd.out<-read.delim("sep.smoothedDD.out.txt",header=T,sep='\t')
#josts d
jostd<-read.delim("p4.jostd.perlocus.txt",header=F)
colnames(jostd)<-c("locid","D")
jostd$SNP<-vcf$SNP
jostd$POS<-vcf$POS
jostd$Chr<-vcf$`#CHROM`
jostd$ID<-vcf$ID
#ld info
ld<-read.delim("ld_info_fwsw.txt")
#neighbor joining SNP trees
fst.trees<-read.delim("p4.snp.ftrees.txt",sep=" ")
#set up some helfpul things
fw.list<-c("TXFW","LAFW","ALFW","FLLG")
sw.list<-c("TXSP","TXCC","TXCB","ALST","FLSG","FLKB",
	"FLFD","FLSI","FLAB","FLPB","FLHB","FLCC")
lgs<-c("LG1","LG2","LG3","LG4","LG5","LG6","LG7","LG8","LG9","LG10","LG11",
	"LG12","LG13","LG14","LG15","LG16","LG17","LG18","LG19","LG20","LG21",
	"LG22")
lgn<-seq(1,22)
all.colors<-c(rep("black",2),"#2166ac","black","#2166ac","black","#2166ac",
        rep("black",8),"#2166ac")
grp.colors<-c('#762a83','#af8dc3','#e7d4e8','#d9f0d3','#7fbf7b','#1b7837')


```

### Define some useful functions

These are updated from **fwsw_analysis.R** and are specific to *these* analyses, and are not generalizable!
```{r functions}
ssc.fwsw.fstsig<-function(tx, la, al, fl, cutoff)
{
  tx.sig<-tx[tx$Fisher.s.P<cutoff,"Locus.ID"]
  la.sig<-la[la$Fisher.s.P<cutoff,"Locus.ID"]
  al.sig<-al[al$Fisher.s.P<cutoff,"Locus.ID"]
  fl.sig<-fl[fl$Fisher.s.P<cutoff,"Locus.ID"]
  length(tx.sig[(tx.sig %in% c(la.sig,al.sig,fl.sig))])
  length(la.sig[(la.sig %in% c(tx.sig,al.sig,fl.sig))])
  length(al.sig[(al.sig %in% c(la.sig,tx.sig,fl.sig))])
  all.shared<-fl.sig[fl.sig %in% la.sig & fl.sig %in% al.sig & fl.sig %in% tx.sig]
  fw.shared.chr<-fwsw.tx[fwsw.tx$Locus.ID %in% all.shared,c("Locus.ID","Chr","BP","Column","Overall.Pi")]
  tapply(fw.shared.chr$Locus.ID,factor(fw.shared.chr$Chr),function(x){ length(unique(x)) })
  #are they using the same SNPs or different SNPs?
  stacks.sig<-data.frame(nrow=nrow(fw.shared.chr),ncol=4)
  for(i in 1:nrow(fw.shared.chr)){
    tx.bp<-fwsw.tx[tx$Fisher.s.P<cutoff & tx$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    la.bp<-fwsw.la[la$Fisher.s.P<cutoff & la$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    al.bp<-fwsw.al[al$Fisher.s.P<cutoff & al$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    fl.bp<-fwsw.fl[fl$Fisher.s.P<cutoff & fl$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    stacks.sig[i,1]<-paste(tx.bp,sep=",",collapse = ",")
    stacks.sig[i,2]<-paste(la.bp,sep=",",collapse = ",")
    stacks.sig[i,3]<-paste(al.bp,sep=",",collapse = ",")
    stacks.sig[i,4]<-paste(fl.bp,sep=",",collapse = ",")
  }
  colnames(stacks.sig)<-c("TX","LA","AL","FL")
  stacks.sig<-data.frame(cbind(fw.shared.chr,stacks.sig))
  stacks.sig$SNP<-paste(stacks.sig$Chr,stacks.sig$BP,sep=".")
  return(stacks.sig)
}

#' extract the significant regions from the gff file
sig.region.ann<-function(fw.shared.chr,gff)
{
  fw.sig.reg<-do.call(rbind,apply(fw.shared.chr,1,function(sig){
    this.gff<-gff[as.character(gff$seqname) %in% as.character(unlist(sig["Chr"])),]
    this.reg<-this.gff[this.gff$start <= as.numeric(sig["BP"]) & this.gff$end >= as.numeric(sig["BP"]),]
    if(nrow(this.reg) == 0){
      if(as.numeric(sig["BP"])>max(as.numeric(this.gff$end))){
        new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                        region="beyond.last.contig", description=NA,SSCID=NA)
      }else{
        new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                        region=NA,description=NA,SSCID=NA)
      }
    }else{
      if(length(grep("SSCG\\d+",this.reg$attribute))>0){
        geneID<-unique(gsub(".*(SSCG\\d+).*","\\1",this.reg$attribute[grep("SSCG\\d+",this.reg$attribute)]))
        gene<-genome.blast[genome.blast$sscv4_gene_ID %in% geneID,"blastp_hit_description"]
      }else{
        geneID<-NA
        gene<-NA
      }
      new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                      region=paste(this.reg$feature,sep=",",collapse = ","),description=gene,SSCID=geneID)
    }
    return(as.data.frame(new))
  }))
}
```

### Combine putative genes and genomic info

```{r putative_regions, eval=TRUE}
#' extract the significant regions from the gff file
put.reg<-do.call(rbind,apply(put.genes,1,function(gene){
  ssc.gene<-as.character(gene[[6]])
  #there might be multiple matches
  ssc.genes<-unlist(strsplit(ssc.gene,","))
  #for each gene, match to gff
  gene.info<-do.call(rbind,lapply(ssc.genes,function(genename){
    this.gff<-gff[grep(genename,gff$attribute),]
    chrom<-unique(as.character(this.gff$seqname))
    start<-min(this.gff$start)
    stop<-max(this.gff$end)
    return(c(chrom,start,stop))
  }))
  if(as.character(gene[[3]]) == "") { gene[[3]]<-NA }
  new<-data.frame(Gene=rep(gene[[1]],length(gene.info[,1])),
                  Function=rep(gene[[2]],length(gene.info[,1])),
                  Chromsome=rep(gene[[3]],length(gene.info[,1])),
                  Species=rep(gene[[4]],length(gene.info[,1])),
                  Citation=rep(gene[[5]],length(gene.info[,1])),
                  Scovelli_geneID=rep(gene[[6]],length(gene.info[,1])),
                  Notes=rep(gene[[7]],length(gene.info[,1])),
                  Chrom=gene.info[,1],StartBP=gene.info[,2],StopBP=gene.info[,3],
                  stringsAsFactors = FALSE)
  return(as.data.frame(new))
}))
write.table(put.reg,"putative.gene.regions.tsv",col.names=T,sep='\t')

```
```{r read_putative,echo=FALSE}
put.reg<-read.delim("putative.gene.regions.tsv",header=T)
```
Note that some genes will have multiple rows as they match multiple locations in the genome.

### Do the putative genes contain RAD loci?
```{r rad_putative}
put.reg$rad.loci<-apply(put.reg,1,function(gene){
  rad<-vcf[vcf$`#CHROM` %in% gene[["Chrom"]] & 
           vcf$POS >= as.numeric(as.character(gene[["StartBP"]])) & 
           vcf$POS <= as.numeric(as.character(gene[["StopBP"]])),]  
  if(nrow(rad)>0){ rad.snps<-paste(rad$SNP,sep=",",collapse=",") }
  else { rad.snps<-NA }
  return(rad.snps)
})
```
```{r, echo=FALSE}
#and how many of the genes have RAD loci?
counts<-tapply(put.reg$rad.loci,put.reg$Gene,function(gene){
  cnt<-length(gene[!is.na(gene)])
})
```

About `r length(counts[counts>0])/length(counts)*100`% of the putative genes have RAD loci within them - but only about `r nrow(put.reg[!is.na(put.reg$rad.loci),])/nrow(put.reg)*100`% of all of the annotations have RAD loci.



## Compare the Stacks outliers with putative genes

### What is the maximum distance of loci in LD?

$D'$ was calculated using a custom C++ program (found at https://github.com/spflanagan/SCA/tree/master/programs/pairwise_ld_vcf) using the formula:

$$D' = \sum_{i=1}^m\sum_{j=1}^n \frac{p_iq_j|D_{ij}|}{D_{max}},$$
where $m$ is the number of alleles at locus and $n$ is the number of alleles at locus B. $D_{ij}$ was calculated as $f_{ij}-p_iq_j$, where $f_{ij}$ is the frequency of the $A_iB_j$ haplotype, $p_i$ is the frequency of allele $i$, and $q_j$ is the frequency of allele $j$. $D_{max}$ is the lesser of $p_iq_j$ or $(1-p_i)(1-q_j)$ when $D_{ij} \lt 0$ and is the minimum of $(1-p_i)q_i$ or $p_i(1-q_i)$ when $D_{ij} \gt 0$.

In the data.frame `ld`, the locus IDs (`LocIDA` and `LocIDB`) are the positions on the chromosome.

```{r ld}
ld$pos.diff<-abs(ld$LocIDA - ld$LocIDB)
chrom.ld<-tapply(ld[ld$D.>0.5,"pos.diff"],ld[ld$D. >0.5,"Chrom"],mean)
```

### Functions for comparison

```{r comp_functions}
#' Find outliers in a putative gene region
outlier.in.region<-function(outlier.df, region.df, oChrom="Chrom", oBP="BP",
                            rChrom="Chrom",rBPStart="StartBP",rBPStop="StopBP"){
  out.in<-apply(region.df,1,function(put.gene){
    oin<-outlier.df[outlier.df[[oChrom]] %in% put.gene[[rChrom]] 
      & outlier.df[[oBP]] >= as.numeric(as.character(put.gene[[rBPStart]]))
      & outlier.df[[oBP]] <= as.numeric(as.character(put.gene[[rBPStop]])),"SNP"]
    if(length(oin)==0){ oin<-NA }
    return(paste(oin,sep=",",collapse=","))
  })
  return(out.in)
}
outlier.nearby<-function(outlier.df,region.df,ld.df,
                         oChrom="Chrom", oBP="BP",
                         rChrom="Chrom",rBPStart="StartBP",rBPStop="StopBP"){
  out.nearby<-apply(region.df,1,function(put.gene){
    nearby<-outlier.df[outlier.df[[oChrom]] %in% put.gene[[rChrom]] 
      & outlier.df[[oBP]] >= (as.numeric(as.character(put.gene[[rBPStart]])) - (ld.df[put.gene[[rChrom]] ]))
      & outlier.df[[oBP]] <= (as.numeric(as.character(put.gene[[rBPStop]])) + (ld.df[put.gene[[rChrom]] ])),"SNP"]
    if(length(nearby)==0){ nearby<-NA }
    return(paste(nearby,sep=",",collapse=","))
  })
}



```

### Prep regions for plotting

```{r fst_plot_prep}

#' Get the plotting positions for the putative genes
all.chr<-data.frame(Chr=vcf$`#CHROM`,BP=vcf$POS,stringsAsFactors = F)
bounds<-data.frame(levels(as.factor(all.chr$Chr)),tapply(as.numeric(as.character(all.chr$BP)),all.chr$Chr,max))
colnames(bounds)<-c("Chrom","End")
plot.scaffs<-scaffs[scaffs %in% bounds$Chrom]

#convert put.reg to the plotting BP
new.dat<-data.frame(stringsAsFactors = F)
last.max<-0
for(i in 1:length(plot.scaffs)){
  #pull out the data for this scaffold
  if(nrow(bounds[bounds$Chrom %in% plot.scaffs[i],])>0){ #sanity check
    chrom.dat<-put.reg[put.reg$Chrom %in% plot.scaffs[i],]
    if(nrow(chrom.dat)>0){
      chrom.dat$plot.min<-as.numeric(as.character(chrom.dat$StartBP))+last.max
      chrom.dat$plot.max<-as.numeric(as.character(chrom.dat$StopBP))+last.max
      new.dat<-rbind(new.dat,chrom.dat)
      #last.max<-max(chrom.dat$plot.pos)+
      #               as.numeric(scaffold.widths[scaffold.widths[,1] %in% scaffs.to.plot[i],2])
    }
    last.max<-last.max+
      as.numeric(bounds[bounds$Chrom %in% plot.scaffs[i],2])
  }else{
    print(paste(plot.scaffs[i], "has no designated width."))
  }
}
#make sure everything is the correct class
new.dat$plot.min<-as.numeric(as.character(new.dat$plot.min))
new.dat$plot.max<-as.numeric(as.character(new.dat$plot.max))
```


### Plot the Fsts 

```{r fst_plot, echo=FALSE,results='hide', message=F,fig.cap="Outlined points are Stacks shared outliers. Black bars are putative genes"}
tx.sig<-fwsw.tx[fwsw.tx$Fisher.s.P<0.01,"Locus.ID"]
la.sig<-fwsw.la[fwsw.la$Fisher.s.P<0.01,"Locus.ID"]
al.sig<-fwsw.al[fwsw.al$Fisher.s.P<0.01,"Locus.ID"]
fl.sig<-fwsw.fl[fwsw.fl$Fisher.s.P<0.01,"Locus.ID"]
length(tx.sig[(tx.sig %in% c(la.sig,al.sig,fl.sig))])
length(la.sig[(la.sig %in% c(tx.sig,al.sig,fl.sig))])
length(al.sig[(al.sig %in% c(la.sig,tx.sig,fl.sig))])
all.shared<-fl.sig[fl.sig %in% la.sig & fl.sig %in% al.sig & fl.sig %in% tx.sig]

par(mfrow=c(4,1),mar=c(0.85,2,0,0.5),oma=c(1,1,1,0.5))
fwswt.fst<-fst.plot(fwsw.tx,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols = c(grp.colors[1],grp.colors[2]),pt.cex=1,axis.size = 1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswt.fst$plot.pos[fwswt.fst$Locus.ID %in% all.shared],fwswt.fst$Corrected.AMOVA.Fst[fwswt.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("TXFW vs. TXCC",3,cex=0.75,line=-1)
fwswl.fst<-fst.plot(fwsw.la,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[2],grp.colors[3]),pt.cex=1,axis.size=1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswl.fst$plot.pos[fwswl.fst$Locus.ID %in% all.shared],fwswl.fst$Corrected.AMOVA.Fst[fwswl.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("LAFW vs. ALST",3,cex=0.75,line=-1)
fwswa.fst<-fst.plot(fwsw.al,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[3],grp.colors[2]),pt.cex=1,axis.size = 1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswa.fst$plot.pos[fwswa.fst$Locus.ID %in% all.shared],fwswa.fst$Corrected.AMOVA.Fst[fwswa.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("ALFW vs. ALST",3,cex=0.75,line=-1)
fwswf.fst<-fst.plot(fwsw.fl,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths =  bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[6],grp.colors[5]),pt.cex=1,axis.size=1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswf.fst$plot.pos[fwswf.fst$Locus.ID %in% all.shared],fwswf.fst$Corrected.AMOVA.Fst[fwswf.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("FLFW vs. FLCC",3,cex=0.75,line=-1)
mtext(expression(italic(F)[ST]),2,outer=T,line=-0.8,cex=0.75)
last<-0
for(i in 1:length(lgs)){
  text(x=mean(fwswf.fst[fwswf.fst$Chr ==lgs[i],"plot.pos"]),y=-0.05,
       labels=lgn[i], adj=1, xpd=TRUE)
  last<-max(fwswf.fst[fwswf.fst$Chr ==lgs[i],"plot.pos"])
}
```

Are outliers in putative gene regions?

```{r fst_compare}
all.put.ssc<-as.character(put.reg$Scovelli_geneID)
all.put.ssc<-unlist(lapply(all.put.ssc,strsplit,","))
sig.put.match<-all.put.ssc[all.put.ssc %in% fw.sig.reg$SSCID]
print(dim(sig.put.match))

#or use the functions
fw.sig.reg$SNP<-paste(fw.sig.reg$Chr,as.character(fw.sig.reg$BP),sep=".")
stacks.in<-outlier.in.region(fw.sig.reg,put.reg,oChrom="Chr")
```

Of the `r nrow(fw.sig.reg)` shared outlier SNPs, `r nrow(fw.sig.reg[!is.na(fw.sig.reg$SSCID),])` are in coding regions of the genome. However, `r length(stacks.in[stacks.in!="NA"])` shared outliers are in putative gene regions.

No, they don't overlap. Maybe they're in the same LD region?

```{r fst_nearby}

put.nearby.rad<-outlier.nearby(fw.sig.reg,put.reg,chrom.ld,oChrom = "Chr")

#add the nearby significant rad loci to the put.reg data.frame
put.reg$nearby.rad<-put.nearby.rad
head(put.reg[put.reg$nearby.rad!="NA",])

```
So of the `r nrow(put.reg)` gene annotations `r length(put.nearby.rad[put.nearby.rad!="NA"])` are within the mean LD range of an outlier, which represent `r length(unique(put.reg[put.reg$nearby.rad!="NA","Gene"]))` putative freshwater genes. Those genes are: `r unique(put.reg[put.reg$nearby.rad!="NA","Gene"])[order(unique(put.reg[put.reg$nearby.rad!="NA","Gene"]))]`

## Compare $\delta$ -divergence to putative regions

Now for $\delta$ -divergence

```{r dd_plot, echo=FALSE, results='hide',fig.keep='all'}
par(mar=c(1,1,1,1),oma=c(1,2,1,1))
dd<-fst.plot(fst.dat = deltad,fst.name = "deltad",bp.name = "Pos",axis=1,pch=19,scaffs.to.plot=scaffs)
mtext(expression(paste(delta,"-divergence")),2,line=1.5)
smooth.par<-data.frame()
smooth.div<-data.frame()
for(i in 1:length(lgs)){#scaffolds are too short
  this.chrom<-dd[dd$Chrom %in% lgs[i],]
  #span<-nrow(this.chrom)/5000
  this.smooth<-loess.smooth(this.chrom$plot.pos,this.chrom$deltad,span=0.1,degree=2) 
  this.plot.pos<-unlist(lapply(this.smooth$x,function(x) { this.chrom[which.min(abs(this.chrom$plot.pos-x)),"plot.pos"] }))
  points(this.smooth$x,this.smooth$y,col="cornflowerblue",type="l",lwd=2)
  this.div<-cbind(dd=this.chrom$deltad[this.chrom$deltad>=quantile(this.chrom$deltad,0.95)],#choosing the upper outliers
                  plot.pos=this.chrom$plot.pos[this.chrom$deltad>=quantile(this.chrom$deltad,0.95)])
  this.par<-cbind(dd=this.chrom$deltad[this.chrom$deltad<=quantile(this.chrom$deltad,0.05)],#choosing the lower outliers
                  plot.pos=this.chrom$plot.pos[this.chrom$deltad<=quantile(this.chrom$deltad,0.05)])
  smooth.par<-rbind(smooth.par,this.par)
  smooth.div<-rbind(smooth.div,this.div)
}
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
```

### Compare $\delta$ -divergence and putative regions

```{r dd_compare}
#in the gene
sdd.in<-outlier.in.region(sdd.out,put.reg,oBP = "Pos")
put.reg$sdd.in<-sdd.in
```

The genes `r unique(put.reg[put.reg$sdd.in!="NA","Gene"])` contain $\delta$-divergence outlier regions.

```{r dd_nearby}
#in the gene
sdd.nb<-outlier.nearby(sdd.out,put.reg,chrom.ld,oBP = "Pos")
```

And of the `r nrow(sdd.out)` $\delta$-divergence outliers, `r length(sdd.nb[sdd.nb!="NA"])/nrow(put.reg)*100`% are in the LD region around putative freshwater genes.


## Compare Bayenv outliers to putative regions

```{r bayenv}
#taken directly from fwsw_analysis.R
bf<-read.delim("bayenv/p4.bf.txt",header=T)
bf$SNP<-paste(bf$scaffold,as.numeric(as.character(bf$BP))+1,sep=".")
bf.co<-apply(bf[,5:7],2,quantile,0.99) #focus on Bayes Factors, because of Lotterhos & Whitlock (2015)
temp.bf.sig<-bf[bf$Temp_BF>bf.co["Temp_BF"],c(1,2,4,8,5,9)]
sal.bf.sig<-bf[bf$Salinity_BF>bf.co["Salinity_BF"],c(1,2,4,8,6,9)]
grass.bf.sig<-bf[bf$seagrass_BF>bf.co["seagrass_BF"],c(1,2,4,8,7,9)]
#get the log transformed Bayes Factors
bf$logSal<-log(bf$Salinity_BF)
bf$logTemp<-log(bf$Temp_BF)
bf$logSeagrass<-log(bf$seagrass_BF)

```

There are `r nrow(temp.bf.sig[temp.bf.sig$locus %in% sal.bf.sig$locus & temp.bf.sig$locus %in% grass.bf.sig,])` overlapping outliers between temperature-, salinity-, and seagrass-associated loci.

But if we only care about salinity ones, there are `r nrow(temp.bf.sig)` outliers.

### Salinity

Are any of the Bayenv salinity outliers near the putative freshwater genes?
```{r bayenv_sal}

bfs.in<-outlier.in.region(sal.bf.sig,put.reg,"scaffold")
bfs.nb<-outlier.nearby(sal.bf.sig,put.reg,chrom.ld,"scaffold")
```
Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bfs.in[bfs.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bfs.nb[bfs.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Temperature

Are any of the Bayenv temperature outliers near putative freshwater genes?

```{r bayenv_temp}

bft.in<-outlier.in.region(temp.bf.sig,put.reg,"scaffold")
bft.nb<-outlier.nearby(temp.bf.sig,put.reg,chrom.ld,"scaffold")
```

Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bft.in[bft.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bft.nb[bft.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Seagrass

Are any of the loci associated with seagrass density in or near putative freshwater genes?

```{r bayenv_grass}

bfg.in<-outlier.in.region(grass.bf.sig,put.reg,"scaffold")
bfg.nb<-outlier.nearby(grass.bf.sig,put.reg,chrom.ld,"scaffold")
```

Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bfg.in[bfg.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bfg.nb[bfg.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Which putative genes contain each of those?

```{r bayenv_genes}
put.reg$bfs.in<-bfs.in
put.reg$bft.in<-bft.in
put.reg$bfg.in<-bfg.in

unique(put.reg[put.reg$bfs.in !="NA","Gene"])
unique(put.reg[put.reg$bft.in !="NA","Gene"])
unique(put.reg[put.reg$bfg.in !="NA","Gene"])
```


All three BayEnv tests identified outliers in the genes `r unique(put.reg[put.reg$bfs.in !="NA","Gene"])[unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bft.in !="NA","Gene"]) & unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bfg.in !="NA","Gene"])]`. The temperature and grass also share `r unique(put.reg[put.reg$bft.in !="NA","Gene"])[unique(put.reg[put.reg$bft.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bfg.in !="NA","Gene"])][1]`, and temperature and salinity share `r unique(put.reg[put.reg$bfs.in !="NA","Gene"])[unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bft.in !="NA","Gene"])][1]`. 

Clearly, RHOGTP8 and TNS1 are showing up repeatedly (all Bayenv2 outliers, and $\delta-divergence outliers).

But looking at `put.reg[put.reg$Gene =="RHOGTP8","rad.loci"]`, it's obvious that Rho GTPase-activating proteins are distributed throughout the genome (on chromosomes `r unique(put.reg[put.reg$Gene=="RHOGTP8","Chrom"])`).

## Neighbor-joining Fst trees

Using `ape`, I calculated neighbor-joining trees and identified `r dim(fst.trees[fst.trees$Monophyletic==TRUE,])` loci with a monophyletic freshwater clade using `is.monophyletic()`. Are any of those SNPs within genes?

```{r nj_trees}
ftmono<-fst.trees[fst.trees$FWMonophyletic == TRUE & !is.na(fst.trees$FWMonophyletic),]

nj.in<-outlier.in.region(ftmono,put.reg,oBP="Pos")
nj.nb<-outlier.nearby(ftmono,put.reg,chrom.ld,oBP="Pos")
put.reg$nj.nb<-nj.nb
put.reg$nj.in<-nj.in
```

So `r length(nj.in[nj.in!="NA"])` loci are located within putative freshwater genes, but `r length(nj.nb[nj.nb!="NA"])` are within the LD region of the putative genes, representing `r length(nj.nb[nj.nb!="NA"])/nrow(put.reg)*100` of the putative gene annotations and `r length(unique(put.reg[put.reg$nj.nb != "NA", "Gene"]))` genes

**I've removed this analysis from the paper, though, so we can ignore this**

## Jost's D

Jost's D measures the fraction of allelic differentiation between populations. According to Molecular Ecologist, ["if allelic differentiation at a particular locus is the value of interest, it appears that D is the best measure"](http://www.molecularecologist.com/2011/03/should-i-use-fst-gst-or-d-2/), so it might be useful to look at Jost's D in the putatively freshwater genes. 

```{r josts_d}
jost.in<-outlier.in.region(jostd,put.reg,oChrom="Chr",oBP="POS")
jost.nb<-outlier.nearby(jostd,put.reg,chrom.ld,oChrom = "Chr",oBP="POS")
put.reg$jostd.nb<-jost.nb
put.reg$jostd.in<-jost.in
```
So `r length(jost.in[jost.in!="NA"])` loci are located within putative freshwater genes, but `r length(jost.nb[jost.nb!="NA"])` are within the LD region of the putative genes, representing `r length(jost.nb[jost.nb!="NA"])/nrow(put.reg)*100` of the putative gene annotations and `r length(unique(put.reg[put.reg$jostd.nb != "NA", "Gene"]))` genes

## Combining all of those analyses

*Genes containing shared Fst outliers: `r unique(put.reg[put.reg$rad.loci != "NA","Gene"])`

*Genes containing $\delta$ -divergence outliers: `r unique(put.reg[put.reg$sdd.in != "NA","Gene"])`

*Genes containing Salinity-associated outliers: `r unique(put.reg[put.reg$bfs.in != "NA","Gene"])`

*Genes containing Jost D outliers: `r unique(put.reg[put.reg$jostd.in !="NA", "Gene"])`

*Genes containing monophyletic FW gene trees [IGNORE]: `r unique(put.reg[put.reg$nj.in != "NA","Gene"])` 

*Overlap: `r unique(put.reg[put.reg$rad.loci != "NA","Gene"])[unique(put.reg[put.reg$rad.loci != "NA","Gene"]) %in% unique(put.reg[put.reg$sdd.in != "NA","Gene"]) &unique(put.reg[put.reg$rad.loci != "NA","Gene"]) %in% unique(put.reg[put.reg$bfs.in != "NA","Gene"])]`

*Divergence analyses: `r unique(put.reg[put.reg$rad.loci != "NA","Gene"])[unique(put.reg[put.reg$rad.loci != "NA","Gene"]) %in% unique(put.reg[put.reg$sdd.in != "NA","Gene"]) &unique(put.reg[put.reg$rad.loci != "NA","Gene"]) %in% unique(put.reg[put.reg$jostd.in != "NA","Gene"])]` 


I'm highlighting a few of the putative genes that have a bunch of outliers nearby or in them. First is TNS1, which matched three genome annotations but only had one region, on LG1, that contained shared outlier Fsts, delta divergence, Bayenv salinity, were near monophyletic neighbor joining trees. 

```{r combine_fsts}
fst.dfs<-list(fwsw.tx,fwsw.la,fwsw.al,fwsw.fl)
names(fst.dfs)<-c("TX FWSW","LA FWSW","AL FWSW","FL FWSW")
colors=c(grp.colors[1],grp.colors[2],grp.colors[3],grp.colors[6])
alphas=c(0.5,0.5,1,0.5)
ltys=c(1,1,2,1)
```

### Plotting function

The plotting function `gene.region.plot` accpets a number of parameters.

```{r plotting_function}
gene.region.plot<-function(chrom,gene,put.reg,vcf,chrom.ld, fst.dfs,deltad=FALSE,D=FALSE,
                           colors="black",lwds=2,alphas=0.5,ltys=1,legend=TRUE,bases="bp", 
                           smooth=FALSE, fst.name="Corrected.AMOVA.FST",txt.cex=1,y.lim=c(0,1),...){
  pstart<-min(as.numeric(as.character(put.reg[put.reg$Gene ==gene & 
                                                     put.reg$Chrom==chrom,"StartBP"])))-(chrom.ld[chrom]/2)
  pstop<-max(as.numeric(as.character(put.reg[put.reg$Gene ==gene & 
                                                    put.reg$Chrom==chrom,"StartBP"])))+(chrom.ld[chrom]/2)
  if(pstart < 0){ pstart<- 0 }
  if(pstop > max(vcf[vcf$`#CHROM`==chrom,"POS"])){ pstop <- max(vcf[vcf$`#CHROM`==chrom,"POS"]) }
  starts<-as.numeric(put.reg[put.reg$Chrom %in% chrom & put.reg$Gene == gene,"StartBP"])
  stops<-as.numeric(put.reg[put.reg$Chrom %in% chrom & put.reg$Gene == gene,"StopBP"])
  
  if(smooth==TRUE){
    #generate the dataframes
    smooth.fsts<-lapply(fst.dfs,function(df){
      this.df<-df[df$Chr %in% chrom,]
      this.smooth<-do.call("rbind",lapply(seq(1,nrow(this.df),(nrow(this.df)*0.15)/5),sliding.avg,
                                        dat=data.frame(Pos=this.df$BP,Fst=this.df[,fst.name]),
                                        width=nrow(this.df)*0.15))
      return(this.smooth)
    })
  }  else{
    smooth.fsts<-lapply(fst.dfs,function(df){
      this.smooth<-df[df$Chr %in% chrom,c("BP",fst.name)]
    })
  }
  
  if(is.data.frame(deltad)){
    this.dd<-deltad[deltad$Chrom %in% chrom,]
    dd.smooth<-loess.smooth(this.dd$Pos,this.dd$deltad,span=0.1,degree=2)
  }
  
  if(is.data.frame(D)){
    this.d<-D[D$Chr %in% chrom,]
    dsmooth<-loess.smooth(this.d$POS,this.d$D,span=0.1,degree=2) 
  }
  pr.gene<-put.reg[put.reg$Chrom==chrom & put.reg$Gene==gene,]
  fst.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$rad.loci[pr.gene$rad.loci != "NA"]),","))
  sal.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$bfs.in[pr.gene$bfs.in != "NA"]),","))
 # tmp.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$bft.in[pr.gene$bft.in != "NA"]),","))
  #grs.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$bfg.in[pr.gene$bfg.in != "NA"]),","))
  #jd.gene<-unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$jostd.nb[jostd.nb != "NA"]),","))
  sdd.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$sdd.in[pr.gene$sdd.in != "NA"]),","))
 #nj.gene<-as.numeric(unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$nj.nb[pr.gene$nj.nb != "NA"]),",")))
  nj.gene<-NULL
  nb.gene<-as.numeric(unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$nearby.rad[pr.gene$nearby.rad != "NA"]),",")))
  gene.out<-data.frame(BP=c(as.numeric(fst.gene),as.numeric(sal.gene),as.numeric(sdd.gene)),
                       shape=c(rep(8,length(fst.gene)),rep(1,length(sal.gene)),rep(4,length(sdd.gene))))
  
  if(length(colors) != length(smooth.fsts)){
    colors<-rep(colors,length(smooth.fsts)/length(colors))
  }
  if(length(lwds) != length(smooth.fsts)){
    lwds<-rep(lwds,length(smooth.fsts)/length(lwds))
  }
  if(length(alphas) != length(smooth.fsts)){
    alphas<-rep(alphas,length(smooth.fsts)/length(alphas))
  }
  if(length(ltys) != length(smooth.fsts)){
    ltys<-rep(ltys,length(smooth.fsts)/length(ltys))
  }
  
  plot(smooth.fsts[[1]],type='n',ylim=c(y.lim[1],y.lim[2]+0.02),
       bty="L",xlab="",ylab="",xaxt='n',yaxt='n',xlim=c(pstart,pstop))
  axis(2,las=1,cex.axis=txt.cex)  
  xlabs<-c(pstart,pstop)
  if(bases %in% c("mb","MB","Mb")) xlabs<-c(round((pstart/1000000),2),round((pstop/1000000),2))
  if(bases %in% c("kb","KB","Kb")) xlabs<-c(round((pstart/1000),2),round((pstop/1000),2))
  axis(1,at=c(pstart,pstop),cex.axis=txt.cex,labels = xlabs)
  #add putative gene
  rect(xleft=as.numeric(starts),xright=as.numeric(stops),
         ybottom=y.lim[1]-0.04,ytop=y.lim[2],col="indianred",border="indianred")
  #add fsts
  mtext(chrom,1,cex=0.75*txt.cex,line = 1)
  for(i in 1:length(smooth.fsts)){
    points(smooth.fsts[[i]],col=alpha(colors[i],alphas[i]),type="l",lwd=lwds[i],lty=ltys[i])
  }
  
  #add delta divergence
  if(is.data.frame(deltad)){
    points(dd.smooth$x,dd.smooth$y,col="#dfc27d",type="l",lwd=2)
  }
  
  #add Jost's D
  if(is.data.frame(D)){
    points(dsmooth$x,dsmooth$y,type="l",col="#a6611a",lwd=2)
  }
  
  
  #are nj trees shown in scope of fig?
  njs.eval<-lapply(nj.gene,function(x){
    if(x >= pstart & x <= pstop) { eval = TRUE }
    else { eval = FALSE }
    return(eval)
  })
  
  #add delta d
  if(is.data.frame(deltad)){
    lgnd<-unlist(lapply(names(fst.dfs),function(n) { bquote(.(n)~italic(F[ST])) }))
    lgnd<-c(lgnd,expression(delta~-divergence))
  }else{
    lgnd<-unlist(lapply(names(fst.dfs),function(n) { substitute(paste(name,italic(F[ST]),sep=""),list(name=n)) }))

  }
  
  cols<-NULL
  pchs<-rep(32,length(lwds))
  for(i in 1:length(colors)){
    cols<-c(cols,alpha(colors[i],alphas[i]))
  }
  if(is.data.frame(deltad)){
    cols<-c(cols,"#dfc27d")
    pchs<-c(pchs,32)
    ltys<-c(ltys,1)
    lwds<-c(lwds,2)
  }
  if(is.data.frame(D)){
    cols<-c(cols,"#a6611a")
    pchs<-c(pchs,32)
    lgnd<-c(lgnd,"Jost's D")
    ltys<-c(ltys,1)
    lwds<-c(lwds,2)
  }
  
  lgnd<-c(lgnd,substitute(italic(g),list(g=gene)),"Outlier RAD loci")
  
  cols<-c(cols,"indianred","black")
  pchs<-c(pchs,15,124)
  if(length(grep(TRUE,njs.eval))>0){
    lgnd<-c(lgnd,"Monophyletic Gene Tree")
    cols<-c(cols,"#08519c")
    pchs<-c(pchs,124)
  }
  
  if(length(sal.gene)>0){
    lgnd<-c(lgnd,"Salinity-associated")
    cols<-c(cols,"black")
    pchs<-c(pchs,25)
  }
  if(legend==TRUE){
  legend("topleft",
         legend=lgnd,pch=pchs,
         bty='n',lwd=c(lwds,0,0,0,0),lty=c(ltys,0,0,0,0),
         col=cols,cex = txt.cex)
  }
  #add outlier loci
  clip(x1=min(as.numeric(pstart)),x2=max(as.numeric(pstop)),y1=y.lim[2]+.01,y2=y.lim[2]+0.2)
  abline(v=fst.gene,col="black")
  points(x=sal.gene,y=rep(y.lim[2]+0.05,length(sal.gene)),col="black",pch=25,cex=0.5*txt.cex)
  abline(v=nb.gene,col=alpha("black",0.5),cex=2)
  abline(v=nj.gene,col="#08519c",cex=2)
}
```

## Plot all genes - for supplement

```{r plot_putative_regions}

par(oma=c(2,2,2,2),mar=c(2,2,2,2))
tns1<-lapply(unique(put.reg[put.reg$Gene=="TNS1" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="TNS1",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst",txt.cex=1.3)

par(mfrow=c(2,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
arhgef3<-lapply(unique(put.reg[put.reg$Gene=="ARHGEF3" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="ARHGEF3",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")

par(mfrow=c(2,4),oma=c(2,2,2,2),mar=c(2,2,2,2))
ca<-lapply(unique(put.reg[put.reg$Gene=="CA" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="CA",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")
ca2<-lapply(unique(put.reg[put.reg$Gene=="CAII" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="CAII",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")

par(mfrow=c(2,4),oma=c(2,2,2,2),mar=c(2,2,2,2))
rhogtp8<-lapply(unique(put.reg[put.reg$Gene=="RHOGTP8" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="RHOGTP8",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")


par(oma=c(2,2,2,2),mar=c(2,2,2,2))
slc26a3<-lapply(unique(put.reg[put.reg$Gene=="SLC26A3" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="SLC26A3",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst") #LG20 doesn't plot

par(mfrow=c(2,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
trim14<-lapply(unique(put.reg[put.reg$Gene=="TRIM14" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="TRIM14",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")

par(oma=c(2,2,2,2),mar=c(2,2,2,2))
pmca<-lapply(unique(put.reg[put.reg$Gene=="PMCA" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="PMCA",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")

par(mfrow=c(1,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
vatpase<-lapply(unique(put.reg[put.reg$Gene=="VATPase" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="VATPase",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst",txt.cex=1.3)
```


```{r old_plots,eval=FALSE,echo=FALSE}
#ones that were important before but not anymore
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
nakatpase<-lapply(unique(put.reg[put.reg$Gene=="NAKATPase" & !is.na(put.reg$rad.loci),"Chrom"]),
       gene.region.plot,gene="NAKATPase",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,D=jostd,colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")


par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
gene.region.plot("LG2","mucin",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                          D=jostd, colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")
gene.region.plot("LG8","mucin",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                        D=jostd,   colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")

par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
gene.region.plot("LG1","CAMKK1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                         D=jostd,  colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")
gene.region.plot("LG11","CAMKK1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                         D=jostd,  colors=colors,alphas=alphas,ltys=ltys,
       smooth=FALSE,fst.name="Smoothed.AMOVA.Fst")
```

### Create figures

Based on the plots above, I'll exclude cases where the candidate gene is beyond the scope of the FST data.

```{r print_putative_regions, eval=FALSE,echo=FALSE}
put.reg$plot<-FALSE
put.reg$plot[!is.na(put.reg$rad.loci)]<-TRUE # | !is.na(put.reg$bfs.in) | !is.na(put.reg$nj.nb)
fave.lgs<-c("LG1","LG2","LG8","LG9","LG12","LG15","LG19")

unique(put.reg[put.reg$Gene=="AQP3","Chrom"])

png("tns1_aqp3.png",height=5,width=10,res=300,units="in")
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
gene.region.plot("LG1","TNS1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
gene.region.plot("LG9","AQP3",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
dev.off()

#PLOT ALL GENES
png("tns1.png",height=10/3,width=10,units="in",res=300)
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
tns1<-lapply(unique(put.reg[put.reg$Gene=="TNS1","Chrom"]),
       gene.region.plot,gene="TNS1",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys)
dev.off()

png("aqp3.png",height=10,width=10,units="in",res=300)
par(mfrow=c(3,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
tns1<-lapply(unique(put.reg[put.reg$Gene=="AQP3" & put.reg$Chrom %in% lgs & put.reg$Chrom != "LG18","Chrom"]),
       gene.region.plot,gene="AQP3",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=3,x.intersp=0)
dev.off()

png("arhgef3.png",height=10,width=10,units="in",res=300)
par(mfrow=c(4,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
arhgef3<-lapply(unique(put.reg[put.reg$Gene=="ARHGEF3" & put.reg$Chrom %in% lgs
                               &!(put.reg$Chrom %in% c("LG5","LG12","LG22")),"Chrom"]),
       gene.region.plot,gene="ARHGEF3",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys)
dev.off()

png("nakatpase.png",height=8,width=10,units="in",res=300)
par(mfrow=c(2,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
nakatpase<-lapply(unique(put.reg[put.reg$Gene=="NAKATPase" & put.reg$Chrom %in% lgs & 
                                   put.reg$Chrom != "LG14","Chrom"]),
       gene.region.plot,gene="NAKATPase",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=1,x.intersp=0)
dev.off()

png("ca.png",height=10,width=7.5,units="in",res=300)
par(mfrow=c(4,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
ca2<-lapply(unique(put.reg[put.reg$Gene=="CAII" &put.reg$Chrom %in% lgs,"Chrom"]),
       gene.region.plot,gene="CAII",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=1,x.intersp=0)
dev.off()

png("mucins.png",height=10,width=10,units="in",res=300)
par(mfrow=c(4,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
ca2<-lapply(unique(put.reg[put.reg$Gene=="mucin" &put.reg$Chrom %in% lgs &
                             put.reg$Chrom != "LG18","Chrom"]),
       gene.region.plot,gene="mucin",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=1,x.intersp=0)
dev.off()

png("camkk1.png",height=7,width=10,units="in",res=300)
par(mfrow=c(2,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
ca2<-lapply(unique(put.reg[put.reg$Gene=="CAMKK1" &put.reg$Chrom %in% lgs,"Chrom"]),
       gene.region.plot,gene="CAMKK1",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=1,x.intersp=0)
dev.off()
       
png("special.mucin.png",height=5,width=10,units="in",res=300)
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
gene.region.plot("LG2","mucin",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
gene.region.plot("LG8","mucin",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
dev.off()

png("camkk1_only2.png",height=5,width=10,units="in",res=300)
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
gene.region.plot("LG1","CAMKK1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
gene.region.plot("LG11","CAMKK1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,
                           colors=colors,alphas=alphas,ltys=ltys)
dev.off()


png("rhogtp.png",height=10,width=10,units="in",res=300)
par(mfrow=c(4,4),oma=c(2,2,2,2),mar=c(2,2,2,2))
rhogtp8<-lapply(unique(put.reg[put.reg$Gene=="RHOGTP8" & put.reg$Chrom %in% lgs & 
                                 put.reg$Chrom != "LG18","Chrom"]),
       gene.region.plot,gene="RHOGTP8",put.reg=put.reg,vcf=vcf,chrom.ld=chrom.ld,
       fst.dfs=fst.dfs,deltad=deltad,colors=colors,alphas=alphas,ltys=ltys,legend=FALSE)
par(mfrow=c(1, 1), oma=rep(0, 4), mar=rep(0, 4), new=TRUE)
plot(0:1,0:1,type='n',xlab="",ylab="",axes=F)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence),
                expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "Monophyletic FW Gene Tree",
                "FW Candidate Gene"),
       pch=c(rep(32,5),124,25,124,124),lty=c(1,1,1,1,1,0,0,0,0),lwd=2,
       col=c(colors,"cornflowerblue","black","black","#08519c","indianred"),
       bty='n',ncol=5,cex=0.75,x.intersp=0)
dev.off()
```

```{r making_Fig7}
colors=c(grp.colors[1],grp.colors[2],grp.colors[3],grp.colors[6])
alphas=c(0.5,0.5,1,0.5)
ltys=c(1,1,2,1)

png("Fig7candidateGenes.png",height=10,width=10,units="in",res=300)


par(mfrow=c(4,4),oma=c(2,2,4.5,2),mar=c(1,3,2,1))
#row1
gene.region.plot("LG1","TNS1",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=18000000,y=0.4,expression(italic(TNS1)),xpd=T,cex=2)
gene.region.plot("LG1","PMCA",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=12000000,y=0.4,expression(italic(PMCA)),xpd=T,cex=2)
gene.region.plot("LG1","CAII",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=4500000,y=0.4,expression(italic(CAII)),xpd=T,cex=2)
gene.region.plot("LG2","CAII",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=7500000,y=0.4,expression(italic(CAII)),xpd=T,cex=2)
#row2
gene.region.plot("LG10","CAII",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=3000000,y=0.4,expression(italic(CAII)),xpd=T,cex=2)
gene.region.plot("LG21","CAII",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=450000,y=0.4,expression(italic(CAII)),xpd=T,cex=2)
gene.region.plot("LG2","TRIM14",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=4000000,y=0.4,expression(italic(TRIM14)),xpd=T,cex=2)
gene.region.plot("LG5","TRIM14",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=7000000,y=0.4,expression(italic(TRIM14)),xpd=T,cex=2)
#row3
gene.region.plot("LG10","TRIM14",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=6000000,y=0.4,expression(italic(TRIM14)),xpd=T,cex=2)
gene.region.plot("LG11","TRIM14",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=2500000,y=0.4,expression(italic(TRIM14)),xpd=T,cex=2)
gene.region.plot("LG13","TRIM14",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=1500000,y=0.4,expression(italic(TRIM14)),xpd=T,cex=2)
gene.region.plot("LG1","VATPase",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=6500000,y=0.4,expression(italic(VATPase)),xpd=T,cex=2)
#row4
gene.region.plot("LG6","VATPase",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.5))
text(x=4500000,y=0.4,expression(italic(VATPase)),xpd=T,cex=2)
gene.region.plot("LG8","VATPase",put.reg,vcf,chrom.ld, fst.dfs,deltad=deltad,txt.cex=2,
                 colors=colors,alphas=alphas,ltys=ltys,legend=FALSE,fst.name="Smoothed.AMOVA.Fst",
                 bases = "MB",y.lim = c(0,0.6))
text(x=4500000,y=0.5,expression(italic(VATPase)),xpd=T,cex=2)




par(fig = c(0, 1, 0, 1),oma=rep(0, 4), mar=rep(0, 4), new=TRUE)
plot(0:1.5, 0:1.5, type="n", xlab="", ylab="", axes=FALSE)
legend(x=0.5,y=0.22,c(expression(Shared~italic(F)[ST]~Outliers),
                "Salinity-associated SNPs",
                "FW Candidate Gene"),
       pch=c(124,25,124),lty=c(0,0,0),lwd=2,
       col=c("black","black","indianred"),
       bty='n',ncol=1,cex=2,x.intersp=-0.5,xpd=T)


#add lines to the top
par(fig = c(0, 1, 0, 1),oma=rep(0, 4), mar=rep(0, 4), new=TRUE)
plot(0:1, 0:1, type="n", xlab="", ylab="", axes=FALSE)
legend("top",c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
                expression(AL~FWSW~italic(F)[ST]),
                expression(FL~FWSW~italic(F)[ST]),
                expression(delta~-divergence)),
       pch=rep(32,5),lty=c(1,1,1,1,1),lwd=2,text.width=0.16,
       col=c(grp.colors[1],grp.colors[2],grp.colors[3],grp.colors[6],
          "#dfc27d"),
       bty='n',ncol=3,cex=2,x.intersp=0.5,y.intersp=0.75,xpd=T)
dev.off()
```

![Fig. 7](../fwsw_results/Fig7candidateGenes.png)

### Plot LG4

```{r plot_lg8,eval=FALSE}
#define a few things
chrom<-"LG4"
pchs<-c(rep(32,6),15,124,124)
cols<-c(grp.colors[1],grp.colors[2],grp.colors[3],grp.colors[6],
        "#dfc27d","#a6611a","indianred","black","#08519c")
ltys=c(1,1,2,1,1,1,0,0,0)
lwds<-c(rep(2,6),0,0,0)

#put together the data for LG4
this.dd<-deltad[deltad$Chrom %in% chrom,]
dd.smooth<-loess.smooth(this.dd$Pos,this.dd$deltad,span=0.1,degree=2)
#add Jost's D
this.d<-jostd[jostd$Chr %in% chrom,]
dsmooth<-loess.smooth(this.d$POS,this.d$D,span=0.1,degree=2) 
#get the outliers
pr.gene<-put.reg[put.reg$Chrom==chrom ,]
#rad loci 
fst.gene<- stacks.sig[stacks.sig$Chr==chrom,"BP"]
#none of the salinity-associated RAD loci are in or near genes on LG8, but we could plot them all anyway
sal.gene<-sal.bf.sig[sal.bf.sig$scaffold=="LG8","BP"]
#jost's d - none are in the putative genes
jd.gene<-unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$jost.in[jost.in != "NA"]),","))
#delta divergence outliers- none are in the putative genes
sdd.gene<- unlist(strsplit(gsub(paste(chrom,".",sep=""),"",pr.gene$sdd.in[pr.gene$sdd.in != "NA"]),","))
#put all of them together with their associated shapes
gene.out<-data.frame(BP=c(as.numeric(fst.gene),as.numeric(sal.gene),as.numeric(sdd.gene)),
                     shape=c(rep(8,length(fst.gene)),rep(1,length(sal.gene)),rep(4,length(sdd.gene))))

png("LG4.png",width=10,height=7,units="in",res=300)

nf <- layout(matrix(c(1,1,1,1,
                      2,2,2,2,
                      3,3,3,4), nrow=3, byrow=TRUE))
par(oma=c(1.5,2.5,1,1),mar=c(1.5,1,2,1))

plot(fst.dfs[[1]]$BP[fst.dfs[[1]]$Chr %in% chrom],
     fst.dfs[[1]]$Smoothed.AMOVA.Fst[fst.dfs[[1]]$Chr %in% chrom],
     type='n',ylim=c(0,1),bty="L",xlab="",ylab="",xaxt='n',yaxt='n')
  axis(2,las=1,cex.axis=2)  
  axis(1,cex.axis=2,at=c(0,5000000,10000000,15000000),labels = c(0,5,10,15))
  #add putative gene
  g<-put.reg[put.reg$Chrom %in% chrom,]
  g$Gene<-as.character(g$Gene)
  g$Gene[grep("ATP6",g$Gene)]<-"ATP6V" #standardize the names
  g<-g[,c("Gene","StartBP","StopBP")]
  g<-g[!duplicated(g$StartBP),]
  g<-g[order(g$StartBP),]
  starts<-as.numeric(g[,"StartBP"])
  stops<-as.numeric(g[,"StopBP"])
  rect(xleft=as.numeric(starts),xright=as.numeric(stops),
         ybottom=-0.04,ytop=1,col="indianred",border="indianred")
  #ad the ones that are spaced out
  text(x=g$StartBP[!g$Gene %in% c("SCUBE1","TRIM14")],y=0.5,
       labels=g$Gene[!g$Gene %in% c("SCUBE1","TRIM14")],font=2,srt=90,xpd=T,cex=1.75)
  text(x=g$StartBP[g$Gene == "SCUBE1"]-50000,y=0.8,
       labels=g$Gene[g$Gene =="SCUBE1"],font=2,srt=90,xpd=T,cex=1.75)
  text(x=g$StartBP[g$Gene == "TRIM14"]+50000,y=0.2,
       labels=g$Gene[g$Gene =="TRIM14"],font=2,srt=90,xpd=T,cex=1.75)
  #add fsts
  mtext(chrom,1,cex=2*0.75,line=2.5)
  for(i in 1:length(fst.dfs)){
    points(fst.dfs[[i]]$BP[fst.dfs[[i]]$Chr %in% chrom],
           fst.dfs[[i]]$Smoothed.AMOVA.Fst[fst.dfs[[i]]$Chr %in% chrom],
           col=alpha(colors[i],alphas[i]),type="l",lwd=lwds[i],lty=ltys[i])
  }
  #add delta-d and Jost's D
  points(dd.smooth$x,dd.smooth$y,col="#dfc27d",type="l",lwd=2)
  points(dsmooth$x,dsmooth$y,type="l",col="#a6611a",lwd=2)   
  #add outlier loci
  clip(x1=min(as.numeric(fst.dfs[[1]]$BP[fst.dfs[[1]]$Chr %in% chrom])),
       x2=max(as.numeric(fst.dfs[[1]]$BP[fst.dfs[[1]]$Chr %in% chrom])),y1=0.99,y2=1.06)
  abline(v=fst.gene,col="black",lwd=1.5)
  points(x=sal.gene,y=rep(1.05,length(sal.gene)),col="black",pch=25,cex=1)
  

  
###Zoom-in plot
plot(fst.dfs[[1]]$BP[fst.dfs[[1]]$Chr %in% chrom],
     fst.dfs[[1]]$Smoothed.AMOVA.Fst[fst.dfs[[1]]$Chr %in% chrom],
     type='n',ylim=c(0,1),bty="L",xlab="",ylab="",xaxt='n',yaxt='n',xlim=c(5000000,10000000))
  axis(2,las=1,cex.axis=2)  
  axis(1,cex.axis=2,at=c(5000000,10000000),labels=c(5,10))
  
  #add putative gene
  g<-put.reg[put.reg$Chrom %in% chrom & put.reg$StartBP<10000000,]
  g$Gene<-as.character(g$Gene)
  g$Gene[grep("ATP6",g$Gene)]<-"ATP6V" #standardize the names
  g<-g[,c("Gene","StartBP","StopBP")]
  g<-g[!duplicated(g$StartBP),]
  g<-g[order(g$StartBP),]
  starts<-as.numeric(g[,"StartBP"])
  stops<-as.numeric(g[,"StopBP"])
  rect(xleft=as.numeric(starts),xright=as.numeric(stops),
         ybottom=-0.04,ytop=1,col="indianred",border="indianred")
   text(x=g$StartBP,y=0.6,labels=g$Gene,font=2,srt=90,xpd=T,cex=1.75)
  # text(x=g$StartBP[1]-20000,y=0.6,labels=g$Gene[1],font=2,srt=90,xpd=T,cex=1.75)
  # text(x=g$StartBP[2]-20000,y=0.6,labels=g$Gene[2],font=2,srt=90,xpd=T,cex=1.75)
  # text(x=g$StartBP[3]-10000,y=0.6,labels=g$Gene[3],font=2,srt=90,xpd=T,cex=1.75)
  # text(x=g$StartBP[4]+10000,y=0.6,labels=g$Gene[4],font=2,srt=90,xpd=T,cex=1.75)
  # text(x=g$StartBP[5],y=0.6,labels=g$Gene[5],font=2,srt=90,xpd=T,cex=1.75)
  
  mtext(chrom,1,cex=2*0.75,line=2)
  for(i in 1:length(fst.dfs)){
    points(fst.dfs[[i]]$BP[fst.dfs[[i]]$Chr %in% chrom],
           fst.dfs[[i]]$Smoothed.AMOVA.Fst[fst.dfs[[i]]$Chr %in% chrom],
           col=alpha(colors[i],alphas[i]),type="l",lwd=lwds[i],lty=ltys[i])
  }
  points(dd.smooth$x,dd.smooth$y,col="#dfc27d",type="l",lwd=2)
  points(dsmooth$x,dsmooth$y,type="l",col="#a6611a",lwd=2)
  clip(x1=min(as.numeric(0)),x2=max(as.numeric(3500000)),y1=0.99,y2=1.06)
    abline(v=fst.gene,col="black")
    points(x=sal.gene,y=rep(1.05,length(sal.gene)),col="black",pch=25,cex=1)
    #abline(v=nb.gene,col=alpha("black",0.5),lwd=1.5)

###Recombination rate
load("~/Projects/ssc_map.rda")
#plot LG4
plot(set[["LG4"]]@interpolations$slidingwindow@physicalPositions,
     set[["LG4"]]@interpolations$slidingwindow@rates,type="l",
     ylim=c(-5,25),xlab="",ylab="",xaxt='n',lwd=2)
abline(h=0,lty=2)
axis(1,at=seq(0,1.8*10^7,3*10^6),labels = seq(0,18,3))
mtext("Position on LG4 (Mb)",1,line=2)
mtext("Recombination Rate (cM/Mb)",2,line=2)
lines(set[["LG4"]]@interpolations$spline@physicalPositions,
      set[["LG4"]]@interpolations$spline@rates,type="l",col="blue",lwd=2)
lines(set[["LG4"]]@interpolations$loess@physicalPositions,
      set[["LG4"]]@interpolations$loess@rates,type="l",lwd=2,col="forestgreen")
legend("topleft",bty='n',col=c("black","blue","forestgreen"),lwd=2,c("Sliding Window","Spline","Loess"))

###add legend

lgnd<-unlist(lapply(names(fst.dfs),function(n) { bquote(.(n)~italic(F[ST])) }))
lgnd<-c(lgnd,expression(delta~-divergence),"Jost's D","gene regions",
        expression("Shared"~italic(F[ST])~"outlier"),"Salinity-associated")
pchs<-c(rep(32,6),15,124,25)
cols<-c(grp.colors[1],grp.colors[2],grp.colors[3],grp.colors[6],
        "#dfc27d","#a6611a","indianred","black","black")
ltys=c(1,1,2,1,1,1,0,0,0)
lwds<-c(rep(2,6),0,0,0)
plot(c(0,1),c(0,1),bty='n',type='n',xlab="",ylab="",xaxt='n',yaxt='n')
legend("topright",legend=lgnd,pch=pchs,
       bty='n',lwd=lwds,lty=ltys,
       col=cols,xpd=TRUE,cex=1.5)
dev.off()
```

![LG 4 Figure](../fwsw_results/LG4.png)