---
title: "Putative Freshwater Genes in Pipefish"
author: "Sarah P. Flanagan"
date: "July 25, 2017"
output: 
  html_document:
    fig_caption: true
---


```{r setup}
knitr::opts_knit$set(root.dir='../fwsw_results/')
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")

```

## Freshwater putative genes

This analysis supplements that found in **fwsw_analysis.R**, which runs the actual analysis of the dataset. Here, I compare the Fst outliers and delta-divergence outliers to putative genes identified from other studies of teleosts adapting to freshwater environments.

### Read in all the files
```{r files}
#putative genes
put.genes<-read.delim("putative_genes.txt",header=TRUE,sep='\t')
#genome annotations
gff.name<-list.files(path = "../../scovelli_genome/",pattern="ssc_\\d+.*_chromlevel.gff")
if(length(grep("gz",gff.name))>0){
  gff<-read.delim(gzfile(paste("../../scovelli_genome/",gff.name,sep="")),header=F)
} else{
  gff<-read.delim(paste("../../scovelli_genome/",gff.name,sep=""),header=F)
}
colnames(gff)<-c("seqname","source","feature","start","end","score","strand","frame","attribute")
#Stacks output
vcf<-parse.vcf("stacks/fw-sw_populations/batch_2.vcf")
vcf$SNP<-paste(vcf$`#CHROM`,vcf$POS,sep=".")
scaffs<-levels(as.factor(vcf[,1]))
fwsw.tx<-read.delim("stacks/batch_2.fst_TXCC-TXFW.tsv")
fwsw.la<-read.delim("stacks/batch_2.fst_ALST-LAFW.tsv")
fwsw.al<-read.delim("stacks/batch_2.fst_ALFW-ALST.tsv")
fwsw.fl<-read.delim("stacks/batch_2.fst_FLCC-FLLG.tsv")
#annotated Stacks outliers at 0.01
fw.sig.reg<-read.csv("StacksFWSWOutliers_annotatedByGenome.csv")
#deltad
deltad<-read.delim("deltadivergence.txt")
sdd.out<-read.delim("smoothed.deltad.out.txt",header=T,sep='\t')
#ld info
ld<-read.delim("ld_info_fwsw.txt")
#neighbor joining SNP trees
fst.trees<-read.delim("ftrees.txt",sep=" ")
#set up some helfpul things
fw.list<-c("TXFW","LAFW","ALFW","FLLG")
sw.list<-c("TXSP","TXCC","TXCB","ALST","FLSG","FLKB",
	"FLFD","FLSI","FLAB","FLPB","FLHB","FLCC")
lgs<-c("LG1","LG2","LG3","LG4","LG5","LG6","LG7","LG8","LG9","LG10","LG11",
	"LG12","LG13","LG14","LG15","LG16","LG17","LG18","LG19","LG20","LG21",
	"LG22")
lgn<-seq(1,22)
all.colors<-c(rep("black",2),"#2166ac","black","#2166ac","black","#2166ac",
        rep("black",8),"#2166ac")
grp.colors<-c('#762a83','#af8dc3','#e7d4e8','#d9f0d3','#7fbf7b','#1b7837')


```

### Define some useful functions

These are updated from **fwsw_analysis.R** and are specific to *these* analyses, and are not generalizable!
```{r functions}
ssc.fwsw.fstsig<-function(tx, la, al, fl, cutoff)
{
  tx.sig<-tx[tx$Fisher.s.P<cutoff,"Locus.ID"]
  la.sig<-la[la$Fisher.s.P<cutoff,"Locus.ID"]
  al.sig<-al[al$Fisher.s.P<cutoff,"Locus.ID"]
  fl.sig<-fl[fl$Fisher.s.P<cutoff,"Locus.ID"]
  length(tx.sig[(tx.sig %in% c(la.sig,al.sig,fl.sig))])
  length(la.sig[(la.sig %in% c(tx.sig,al.sig,fl.sig))])
  length(al.sig[(al.sig %in% c(la.sig,tx.sig,fl.sig))])
  all.shared<-fl.sig[fl.sig %in% la.sig & fl.sig %in% al.sig & fl.sig %in% tx.sig]
  fw.shared.chr<-fwsw.tx[fwsw.tx$Locus.ID %in% all.shared,c("Locus.ID","Chr","BP","Column","Overall.Pi")]
  tapply(fw.shared.chr$Locus.ID,factor(fw.shared.chr$Chr),function(x){ length(unique(x)) })
  #are they using the same SNPs or different SNPs?
  stacks.sig<-data.frame(nrow=nrow(fw.shared.chr),ncol=4)
  for(i in 1:nrow(fw.shared.chr)){
    tx.bp<-fwsw.tx[tx$Fisher.s.P<cutoff & tx$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    la.bp<-fwsw.la[la$Fisher.s.P<cutoff & la$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    al.bp<-fwsw.al[al$Fisher.s.P<cutoff & al$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    fl.bp<-fwsw.fl[fl$Fisher.s.P<cutoff & fl$Locus.ID == fw.shared.chr[i,"Locus.ID"],"BP"]
    stacks.sig[i,1]<-paste(tx.bp,sep=",",collapse = ",")
    stacks.sig[i,2]<-paste(la.bp,sep=",",collapse = ",")
    stacks.sig[i,3]<-paste(al.bp,sep=",",collapse = ",")
    stacks.sig[i,4]<-paste(fl.bp,sep=",",collapse = ",")
  }
  colnames(stacks.sig)<-c("TX","LA","AL","FL")
  stacks.sig<-data.frame(cbind(fw.shared.chr,stacks.sig))
  stacks.sig$SNP<-paste(stacks.sig$Chr,stacks.sig$BP,sep=".")
  return(stacks.sig)
}

#' extract the significant regions from the gff file
sig.region.ann<-function(fw.shared.chr,gff)
{
  fw.sig.reg<-do.call(rbind,apply(fw.shared.chr,1,function(sig){
    this.gff<-gff[as.character(gff$seqname) %in% as.character(unlist(sig["Chr"])),]
    this.reg<-this.gff[this.gff$start <= as.numeric(sig["BP"]) & this.gff$end >= as.numeric(sig["BP"]),]
    if(nrow(this.reg) == 0){
      if(as.numeric(sig["BP"])>max(as.numeric(this.gff$end))){
        new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                        region="beyond.last.contig", description=NA,SSCID=NA)
      }else{
        new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                        region=NA,description=NA,SSCID=NA)
      }
    }else{
      if(length(grep("SSCG\\d+",this.reg$attribute))>0){
        geneID<-unique(gsub(".*(SSCG\\d+).*","\\1",this.reg$attribute[grep("SSCG\\d+",this.reg$attribute)]))
        gene<-genome.blast[genome.blast$sscv4_gene_ID %in% geneID,"blastp_hit_description"]
      }else{
        geneID<-NA
        gene<-NA
      }
      new<-data.frame(Locus=sig["Locus.ID"],Chr=sig["Chr"],BP=sig["BP"],SNPCol=sig["Column"],
                      region=paste(this.reg$feature,sep=",",collapse = ","),description=gene,SSCID=geneID)
    }
    return(as.data.frame(new))
  }))
}
```

### Combine putative genes and genomic info

```{r putative_regions}
#' extract the significant regions from the gff file
put.reg<-do.call(rbind,apply(put.genes,1,function(gene){
  ssc.gene<-as.character(gene[[6]])
  #there might be multiple matches
  ssc.genes<-unlist(strsplit(ssc.gene,","))
  #for each gene, match to gff
  gene.info<-do.call(rbind,lapply(ssc.genes,function(genename){
    this.gff<-gff[grep(genename,gff$attribute),]
    chrom<-unique(as.character(this.gff$seqname))
    start<-min(this.gff$start)
    stop<-max(this.gff$end)
    return(c(chrom,start,stop))
  }))
  if(as.character(gene[[3]]) == "") { gene[[3]]<-NA }
  new<-data.frame(Gene=rep(gene[[1]],length(gene.info[,1])),
                  Function=rep(gene[[2]],length(gene.info[,1])),
                  Chromsome=rep(gene[[3]],length(gene.info[,1])),
                  Species=rep(gene[[4]],length(gene.info[,1])),
                  Citation=rep(gene[[5]],length(gene.info[,1])),
                  Scovelli_geneID=rep(gene[[6]],length(gene.info[,1])),
                  Notes=rep(gene[[7]],length(gene.info[,1])),
                  Chrom=gene.info[,1],StartBP=gene.info[,2],StopBP=gene.info[,3],
                  stringsAsFactors = FALSE)
  return(as.data.frame(new))
}))

```

Note that some genes will have multiple rows as they match multiple locations in the genome.

### Do the putative genes contain RAD loci?
```{r rad_putative}
put.reg$rad.loci<-apply(put.reg,1,function(gene){
  rad<-vcf[vcf$`#CHROM` %in% gene[["Chrom"]] & 
           vcf$POS >= as.numeric(as.character(gene[["StartBP"]])) & 
           vcf$POS <= as.numeric(as.character(gene[["StopBP"]])),]  
  if(nrow(rad)>0){ rad.snps<-paste(rad$SNP,sep=",",collapse=",") }
  else { rad.snps<-NA }
  return(rad.snps)
})
```
```{r, echo=FALSE}
#and how many of the genes have RAD loci?
counts<-tapply(put.reg$rad.loci,put.reg$Gene,function(gene){
  cnt<-length(gene[!is.na(gene)])
})
```

About `r length(counts[counts>0])/length(counts)*100`% of the putative genes have RAD loci within them - but only about `r nrow(put.reg[!is.na(put.reg$rad.loci),])/nrow(put.reg)*100`% of all of the annotations have RAD loci.

## Compare the Stacks outliers with putative genes

### What is the maximum distance of loci in LD?

$D'$ was calculated using a custom C++ program (found at https://github.com/spflanagan/SCA/tree/master/programs/pairwise_ld_vcf) using the formula:

$$D' = \sum_{i=1}^m\sum_{j=1}^n \frac{p_iq_j|D_{i,j}|}{D_{max}},$$
where $m$ is the number of alleles at locus and $n$ is the number of alleles at locus B. $D_{ij}$ was calculated as $f_{ij}-p_iq_j$, where $f_{ij}$ is the frequency of the $A_iB_j$ haplotype, $p_i$ is the frequency of allele $i$, and $q_j$ is the frequency of allele $j$. $D_{max}$ is the lesser of $p_iq_j$ or $q-p_iq_j$ when $D_{ij} \lt 0$ and is the minimum of $(1-p_i)q_i$ or $p_i(1-q_i)$ when $D_{ij} \gt 0$.

In the data.frame `ld`, the locus IDs (`LocIDA` and `LocIDB`) are the positions on the chromosome.

```{r ld}
ld$pos.diff<-abs(ld$LocIDA - ld$LocIDB)
chrom.ld<-tapply(ld[ld$D.>0.5,"pos.diff"],ld[ld$D. >0.5,"Chrom"],mean)
```

### Functions for comparison

```{r comp_functions}
#' Find outliers in a putative gene region
outlier.in.region<-function(outlier.df, region.df, oChrom="Chrom", oBP="BP",
                            rChrom="Chrom",rBPStart="StartBP",rBPStop="StopBP"){
  out.in<-apply(region.df,1,function(put.gene){
    oin<-outlier.df[outlier.df[[oChrom]] %in% put.gene[[rChrom]] 
      & outlier.df[[oBP]] >= as.numeric(as.character(put.gene[[rBPStart]]))
      & outlier.df[[oBP]] <= as.numeric(as.character(put.gene[[rBPStop]])),"SNP"]
    if(length(oin)==0){ oin<-NA }
    return(paste(oin,sep=",",collapse=","))
  })
  return(out.in)
}
outlier.nearby<-function(outlier.df,region.df,ld.df,
                         oChrom="Chrom", oBP="BP",
                         rChrom="Chrom",rBPStart="StartBP",rBPStop="StopBP"){
  out.nearby<-apply(region.df,1,function(put.gene){
    nearby<-outlier.df[outlier.df[[oChrom]] %in% put.gene[[rChrom]] 
      & outlier.df[[oBP]] >= (as.numeric(as.character(put.gene[[rBPStart]])) - (ld.df[put.gene[[rChrom]] ]))
      & outlier.df[[oBP]] <= (as.numeric(as.character(put.gene[[rBPStop]])) + (ld.df[put.gene[[rChrom]] ])),"SNP"]
    if(length(nearby)==0){ nearby<-NA }
    return(paste(nearby,sep=",",collapse=","))
  })
}



```

### Prep regions for plotting

```{r fst_plot_prep}

#' Get the plotting positions for the putative genes
all.chr<-data.frame(Chr=vcf$`#CHROM`,BP=vcf$POS,stringsAsFactors = F)
bounds<-data.frame(levels(as.factor(all.chr$Chr)),tapply(as.numeric(as.character(all.chr$BP)),all.chr$Chr,max))
colnames(bounds)<-c("Chrom","End")
plot.scaffs<-scaffs[scaffs %in% bounds$Chrom]

#convert put.reg to the plotting BP
new.dat<-data.frame(stringsAsFactors = F)
last.max<-0
for(i in 1:length(plot.scaffs)){
  #pull out the data for this scaffold
  if(nrow(bounds[bounds$Chrom %in% plot.scaffs[i],])>0){ #sanity check
    chrom.dat<-put.reg[put.reg$Chrom %in% plot.scaffs[i],]
    if(nrow(chrom.dat)>0){
      chrom.dat$plot.min<-as.numeric(as.character(chrom.dat$StartBP))+last.max
      chrom.dat$plot.max<-as.numeric(as.character(chrom.dat$StopBP))+last.max
      new.dat<-rbind(new.dat,chrom.dat)
      #last.max<-max(chrom.dat$plot.pos)+
      #               as.numeric(scaffold.widths[scaffold.widths[,1] %in% scaffs.to.plot[i],2])
    }
    last.max<-last.max+
      as.numeric(bounds[bounds$Chrom %in% plot.scaffs[i],2])
  }else{
    print(paste(plot.scaffs[i], "has no designated width."))
  }
}
#make sure everything is the correct class
new.dat$plot.min<-as.numeric(as.character(new.dat$plot.min))
new.dat$plot.max<-as.numeric(as.character(new.dat$plot.max))
```


### Plot the Fsts 

```{r fst_plot, echo=FALSE, message=F,fig.cap="Outlined points are Stacks shared outliers. Black bars are putative genes"}
tx.sig<-fwsw.tx[fwsw.tx$Fisher.s.P<0.01,"Locus.ID"]
la.sig<-fwsw.la[fwsw.la$Fisher.s.P<0.01,"Locus.ID"]
al.sig<-fwsw.al[fwsw.al$Fisher.s.P<0.01,"Locus.ID"]
fl.sig<-fwsw.fl[fwsw.fl$Fisher.s.P<0.01,"Locus.ID"]
length(tx.sig[(tx.sig %in% c(la.sig,al.sig,fl.sig))])
length(la.sig[(la.sig %in% c(tx.sig,al.sig,fl.sig))])
length(al.sig[(al.sig %in% c(la.sig,tx.sig,fl.sig))])
all.shared<-fl.sig[fl.sig %in% la.sig & fl.sig %in% al.sig & fl.sig %in% tx.sig]

par(mfrow=c(4,1),mar=c(0.85,2,0,0.5),oma=c(1,1,1,0.5))
fwswt.fst<-fst.plot(fwsw.tx,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols = c(grp.colors[1],grp.colors[2]),pt.cex=1,axis.size = 1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswt.fst$plot.pos[fwswt.fst$Locus.ID %in% all.shared],fwswt.fst$Corrected.AMOVA.Fst[fwswt.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("TXFW vs. TXCC",3,cex=0.75,line=-1)
fwswl.fst<-fst.plot(fwsw.la,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[2],grp.colors[3]),pt.cex=1,axis.size=1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswl.fst$plot.pos[fwswl.fst$Locus.ID %in% all.shared],fwswl.fst$Corrected.AMOVA.Fst[fwswl.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("LAFW vs. ALST",3,cex=0.75,line=-1)
fwswa.fst<-fst.plot(fwsw.al,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths = bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[3],grp.colors[2]),pt.cex=1,axis.size = 1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswa.fst$plot.pos[fwswa.fst$Locus.ID %in% all.shared],fwswa.fst$Corrected.AMOVA.Fst[fwswa.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("ALFW vs. ALST",3,cex=0.75,line=-1)
fwswf.fst<-fst.plot(fwsw.fl,fst.name = "Corrected.AMOVA.Fst", bp.name = "BP",chrom.name = "Chr", 
                    scaffs.to.plot=plot.scaffs, scaffold.widths =  bounds,pch=19,y.lim = c(0,1),
                    pt.cols=c(grp.colors[6],grp.colors[5]),pt.cex=1,axis.size=1)
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
points(fwswf.fst$plot.pos[fwswf.fst$Locus.ID %in% all.shared],fwswf.fst$Corrected.AMOVA.Fst[fwswf.fst$Locus.ID %in% all.shared],
       pch=1,col="black",cex=1.3)
mtext("FLFW vs. FLCC",3,cex=0.75,line=-1)
mtext(expression(italic(F)[ST]),2,outer=T,line=-0.8,cex=0.75)
last<-0
for(i in 1:length(lgs)){
  text(x=mean(fwswf.fst[fwswf.fst$Chr ==lgs[i],"plot.pos"]),y=-0.05,
       labels=lgn[i], adj=1, xpd=TRUE)
  last<-max(fwswf.fst[fwswf.fst$Chr ==lgs[i],"plot.pos"])
}
```

They look like they match up in some places, but do they really?

```{r fst_compare}
all.put.ssc<-as.character(put.reg$Scovelli_geneID)
all.put.ssc<-unlist(lapply(all.put.ssc,strsplit,","))
sig.put.match<-all.put.ssc[all.put.ssc %in% fw.sig.reg$SSCID]
print(dim(sig.put.match))

#or use the functions
fw.sig.reg$SNP<-paste(fw.sig.reg$Chr,as.character(fw.sig.reg$BP),sep=".")
stacks.in<-outlier.in.region(fw.sig.reg,put.reg,oChrom="Chr")
```

No, they don't overlap. Maybe they're in the same LD region?

```{r fst_nearby}

put.nearby.rad<-outlier.nearby(fw.sig.reg,put.reg,chrom.ld,oChrom = "Chr")

#add the nearby significant rad loci to the put.reg data.frame
put.reg$nearby.rad<-put.nearby.rad
head(put.reg[put.reg$nearby.rad!="NA",])

```
So of the `r nrow(put.reg)` gene annotations `r length(put.nearby.rad[put.nearby.rad!="NA"])` are within the mean LD range of an outlier, which represent `r length(unique(put.reg[put.reg$nearby.rad!="NA","Gene"]))` putative freshwater genes. Those genes are: `r unique(put.reg[put.reg$nearby.rad!="NA","Gene"])[order(unique(put.reg[put.reg$nearby.rad!="NA","Gene"]))]`

## Compare \delta-divergence to putative regions

Now for \delta-divergence

```{r dd_plot, echo=FALSE, results='hide',fig.keep='all'}
par(mar=c(1,1,1,1),oma=c(1,2,1,1))
dd<-fst.plot(fst.dat = deltad,fst.name = "deltad",bp.name = "Pos",axis=1,pch=19,scaffs.to.plot=scaffs)
mtext(expression(paste(delta,"-divergence")),2,line=1.5)
smooth.par<-data.frame()
smooth.div<-data.frame()
for(i in 1:length(lgs)){#scaffolds are too short
  this.chrom<-dd[dd$Chrom %in% lgs[i],]
  #span<-nrow(this.chrom)/5000
  this.smooth<-loess.smooth(this.chrom$plot.pos,this.chrom$deltad,span=0.1,degree=2) 
  this.plot.pos<-unlist(lapply(this.smooth$x,function(x) { this.chrom[which.min(abs(this.chrom$plot.pos-x)),"plot.pos"] }))
  points(this.smooth$x,this.smooth$y,col="cornflowerblue",type="l",lwd=2)
  this.div<-cbind(dd=this.chrom$deltad[this.chrom$deltad>=quantile(this.chrom$deltad,0.95)],#choosing the upper outliers
                  plot.pos=this.chrom$plot.pos[this.chrom$deltad>=quantile(this.chrom$deltad,0.95)])
  this.par<-cbind(dd=this.chrom$deltad[this.chrom$deltad<=quantile(this.chrom$deltad,0.05)],#choosing the lower outliers
                  plot.pos=this.chrom$plot.pos[this.chrom$deltad<=quantile(this.chrom$deltad,0.05)])
  smooth.par<-rbind(smooth.par,this.par)
  smooth.div<-rbind(smooth.div,this.div)
}
apply(new.dat,1,function(dat){
  arrows(x0=as.numeric(dat[["plot.min"]]),x1=as.numeric(dat[["plot.max"]]),y0=0.9,y1=0.9,length=0,lwd=10)
})
```

### Compare \delta-divergence and putative regions

```{r dd_compare}
#in the gene
sdd.in<-outlier.in.region(sdd.out,put.reg,oBP = "Pos")
put.reg$sdd.in<-sdd.in
```

The genes `r unique(put.reg[put.reg$sdd.in!="NA","Gene"])` contain delta divergence outlier regions.

```{r dd_nearby}
#in the gene
sdd.nb<-outlier.nearby(sdd.out,put.reg,chrom.ld,oBP = "Pos")
```

And of the `r nrow(sdd.out)` \delta-divergence outliers, `r length(sdd.nb[sdd.nb!="NA"])/nrow(put.reg)*100`% are in the LD region around putative freshwater genes.


## Compare Bayenv outliers to putative regions

```{r bayenv}
#taken directly from fwsw_analysis.R
bf<-read.delim("bf.txt",header=T)
bf$SNP<-paste(bf$scaffold,as.numeric(as.character(bf$BP))+1,sep=".")
bf.co<-apply(bf[,5:7],2,quantile,0.99) #focus on Bayes Factors, because of Lotterhos & Whitlock (2015)
temp.bf.sig<-bf[bf$Temp_BF>bf.co["Temp_BF"],c(1,2,4,8,5,9)]
sal.bf.sig<-bf[bf$Salinity_BF>bf.co["Salinity_BF"],c(1,2,4,8,6,9)]
grass.bf.sig<-bf[bf$seagrass_BF>bf.co["seagrass_BF"],c(1,2,4,8,7,9)]
#get the log transformed Bayes Factors
bf$logSal<-log(bf$Salinity_BF)
bf$logTemp<-log(bf$Temp_BF)
bf$logSeagrass<-log(bf$seagrass_BF)

```

There are `r dim(temp.bf.sig[temp.bf.sig$locus %in% sal.bf.sig$locus & temp.bf.sig$locus %in% grass.bf.sig,])` overlapping outliers between temperature-, salinity-, and seagrass-associated loci.

### Salinity

Are any of the Bayenv salinity outliers near the putative freshwater genes?
```{r bayenv_temp}

bfs.in<-outlier.in.region(sal.bf.sig,put.reg,"scaffold")
bfs.nb<-outlier.nearby(sal.bf.sig,put.reg,chrom.ld,"scaffold")
```
Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bfs.in[bfs.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bfs.nb[bfs.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Temperature

Are any of the Bayenv temperature outliers near putative freshwater genes?

```{r bayenv_temp}

bft.in<-outlier.in.region(temp.bf.sig,put.reg,"scaffold")
bft.nb<-outlier.nearby(temp.bf.sig,put.reg,chrom.ld,"scaffold")
```

Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bft.in[bft.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bft.nb[bft.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Seagrass

Are any of the loci associated with seagrass density in or near putative freshwater genes?

```{r bayenv_grass}

bfg.in<-outlier.in.region(grass.bf.sig,put.reg,"scaffold")
bfg.nb<-outlier.nearby(grass.bf.sig,put.reg,chrom.ld,"scaffold")
```

Of the `r nrow(bf)` RAD loci analyzed by *Bayenv2* for associations with temperature, `r length(bfg.in[bfg.in!="NA"])/nrow(put.reg)*100`% are in putative freshwater genes and `r length(bfg.nb[bfg.nb!="NA"])/nrow(put.reg)*100`% are within the LD neighborhood.

### Which putative genes contain each of those?

```{r bayenv_genes}
put.reg$bfs.in<-bfs.in
put.reg$bft.in<-bft.in
put.reg$bfg.in<-bfg.in


unique(put.reg[put.reg$bft.in !="NA","Gene"])
unique(put.reg[put.reg$bfg.in !="NA","Gene"])
```


All three BayEnv tests identified outliers in the genes `r unique(put.reg[put.reg$bfs.in !="NA","Gene"])[unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bft.in !="NA","Gene"]) & unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bfg.in !="NA","Gene"])]`. The temperature and grass also share `r unique(put.reg[put.reg$bft.in !="NA","Gene"])[unique(put.reg[put.reg$bft.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bfg.in !="NA","Gene"])][1]`, and temperature and salinity share `r unique(put.reg[put.reg$bfs.in !="NA","Gene"])[unique(put.reg[put.reg$bfs.in !="NA","Gene"]) %in% unique(put.reg[put.reg$bft.in !="NA","Gene"])][1]`. 

Clearly, RHOGTP8 and TNS1 are showing up repeatedly (all Bayenv2 outliers, and \delta-divergence outliers).

But looking at `put.reg[put.reg$Gene =="RHOGTP8",]`, it's obvious that Rho GTPase-activating proteins are distributed throughout the genome (on chromosomes `r unique(put.reg[put.reg$Gene=="RHOGTP8","Chrom"])`).

## Neighbor-joining Fst trees

Using `ape`, I calculated neighbor-joining trees and identified `r dim(fst.trees[fst.trees$Monophyletic==TRUE,])` loci with a monophyletic freshwater clade using `is.monophyletic()`. Are any of those SNPs within genes?

```{r nj_trees}
ftmono<-fst.trees[fst.trees$Monophyletic == TRUE,]

nj.in<-outlier.in.region(ftmono,put.reg,oBP="Pos")
nj.nb<-outlier.nearby(ftmono,put.reg,chrom.ld,oBP="Pos")
put.reg$nj.nb<-nj.nb
```

So `r length(nj.in[nj.in!="NA"])` loci are located within putative freshwater genes, but `r length(nj.nb[nj.nb!="NA"])` are within the LD region of the putative genes, representing `r length(nj.nb[nj.nb!="NA"])/nrow(put.reg)*100` of the putative gene annotations and `r length(unique(put.reg[put.reg$nj.nb != "NA", "Gene"]))` genes
## Combining all of those analyses
```{r plot_tns1}
tns1.pstart<-min(as.numeric(as.character(put.reg[put.reg$Gene =="TNS1" & 
                                                   put.reg$Chrom=="LG1","StartBP"])))-(chrom.ld["LG1"]/2)
tns1.pstop<-max(as.numeric(as.character(put.reg[put.reg$Gene =="TNS1" & 
                                                   put.reg$Chrom=="LG1","StartBP"])))+(chrom.ld["LG1"]/2)
if(tns1.pstop > max(vcf[vcf$`#CHROM`=="LG1","POS"])){ tns1.pstop <- max(vcf[vcf$`#CHROM`=="LG1","POS"]) }
starts<-as.numeric(put.reg[put.reg$Chrom %in% "LG1" & put.reg$Gene == "TNS1","StartBP"])
stops<-as.numeric(put.reg[put.reg$Chrom %in% "LG1" & put.reg$Gene == "TNS1","StopBP"])

#generate the dataframes
this.tx<-fwsw.tx[fwsw.tx$Chr %in% "LG1",]
tx.smooth<-do.call("rbind",lapply(seq(1,nrow(this.tx),(nrow(this.tx)*0.15)/5),sliding.avg,
                                  dat=data.frame(Pos=this.tx$BP,Fst=this.tx$Corrected.AMOVA.Fst),
                                  width=nrow(this.tx)*0.15))
this.la<-fwsw.la[fwsw.la$Chr %in% "LG1",]
la.smooth<-do.call("rbind",lapply(seq(1,nrow(this.la),(nrow(this.la)*0.15)/5),sliding.avg,
                                  dat=data.frame(Pos=this.la$BP,Fst=this.la$Corrected.AMOVA.Fst),
                                  width=nrow(this.la)*0.15))
this.al<-fwsw.al[fwsw.al$Chr %in% "LG1",]
al.smooth<-do.call("rbind",lapply(seq(1,nrow(this.al),(nrow(this.al)*0.15)/5),sliding.avg,
                                  dat=data.frame(Pos=this.al$BP,Fst=this.al$Corrected.AMOVA.Fst),
                                  width=nrow(this.al)*0.15))
this.fl<-fwsw.fl[fwsw.fl$Chr %in% "LG1",]
fl.smooth<-do.call("rbind",lapply(seq(1,nrow(this.fl),(nrow(this.fl)*0.15)/5),sliding.avg,
                                  dat=data.frame(Pos=this.fl$BP,Fst=this.fl$Corrected.AMOVA.Fst),
                                  width=nrow(this.fl)*0.15))
this.chrom<-deltad[deltad$Chrom %in% "LG1",]
this.smooth<-loess.smooth(this.chrom$Pos,this.chrom$deltad,span=0.1,degree=2)
pr.tns1<-put.reg[put.reg$Chrom=="LG1" & put.reg$Gene=="TNS1",]
sal.tns1<- gsub("LG1.(\\d+)","\\1",pr.tns1$bfs.in[pr.tns1$bfs.in != "NA"])
tmp.tns1<- gsub("LG1.(\\d+)","\\1",pr.tns1$bft.in[pr.tns1$bft.in != "NA"])
tmp.tns1<-unlist(strsplit(tmp.tns1,','))
grs.tns1<- gsub("LG1.(\\d+)","\\1",pr.tns1$bfg.in[pr.tns1$bfg.in != "NA"])
sdd.tns1<- gsub("LG1.(\\d+)","\\1",pr.tns1$sdd.in[pr.tns1$sdd.in != "NA"])
tns1.out<-data.frame(BP=c(as.numeric(sal.tns1),as.numeric(tmp.tns1),as.numeric(grs.tns1),as.numeric(sdd.tns1)),
                     shape=c(rep(1,length(sal.tns1)),rep(2,length(tmp.tns1)),
                             rep(3,length(grs.tns1)),rep(4,length(sdd.tns1))))
#plot
plot(tx.smooth,col=alpha(grp.colors[1],0.5),type="l",ylim=c(0,1),lwd=2,
     bty="L",xlab="",ylab="",xaxt='n',yaxt='n',xlim=c(tns1.pstart,tns1.pstop))
axis(2,las=1)  
mtext("Position on LG1",1)
points(la.smooth,col=alpha(grp.colors[2],0.5),type="l",lwd=2)
points(al.smooth,col=grp.colors[3],type="l",lwd=2,lty=2)
points(fl.smooth,col=alpha(grp.colors[6],0.5),type="l",lwd=2)
#add delta divergence
points(this.smooth$x,this.smooth$y,col="cornflowerblue",type="l",lwd=2)
#add putative genes' LD regions
#arrows(x0=tns1.pstart,x1=tns1.pstop,y0=0.8,y1=0.8,col="grey",lwd=15,length=0)
#add putative genes
arrows(x0=as.numeric(starts),x1=as.numeric(stops),
       y0=0.8,y1=0.8,col="indianred",lwd=15,length=0)
#add outlier loci
points(x = fw.sig.reg[fw.sig.reg$Chr == "LG1","BP"],
        y = rep(0.7,length(fw.sig.reg[fw.sig.reg$Chr == "LG1","BP"])),col="red")#these don't exist
clip(x1=min(as.numeric(starts)),x2=max(as.numeric(stops)),y1=0.78,y2=0.82)
abline(v=tns1.out$BP,col="black")
#clip(x1=0,x2=tns1.pstop,y1=0,y2=1)
mtext(expression(italic(F)[ST]),2,line=2)
legend("topleft",
  legend=c(expression(TX~FWSW~italic(F)[ST]),
                expression(LA~FWSW~italic(F)[ST]),
         expression(AL~FWSW~italic(F)[ST]),expression(FL~FWSW~italic(F)[ST]),
         expression(delta~-divergence),expression(italic(TNS1)),"Outlier RAD loci"),
  bty='n',lwd=2,lty=c(1,1,2,1,1,1,1),
       col=c(alpha(grp.colors[1],0.5),alpha(grp.colors[2],0.5),
            grp.colors[3],alpha(grp.colors[6],0.5),
            "cornflowerblue", "indianred","black"))
```




