---
title: "Run dmc"
author: "Sarah P. Flanagan"
date: "March 5, 2018"
output:
  pdf_document: default
  html_document: default
---

First I'll set up the working environment.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir='../fwsw_results/')
```

```{r source, include=FALSE}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
library(knitr)
library(fields)
library(MASS)
library(RColorBrewer)
```

and load the necessary files
```{r access_analysis, echo=FALSE}
knitr::read_chunk("../scripts/fwsw_analysis.R")
knitr::read_chunk("../scripts/run_dmc.R")
```
```{r FWSWsetup, include=FALSE}
```
```{r read_vcf}
vcf<-parse.vcf("stacks/p4.vcf") #this is the smaller dataset
```
```{r vcfSetup, echo=TRUE}
```

Now I'll go through the steps outline in https://github.com/kristinmlee/dmc/blob/master/dmc_example.md

## Calculate neutral F matrix

For this first step (calculating the neutral variance/covariance matrix, F), I'll use linkage groups that have no or few shared outlier loci. I need to (1) Calculate allele frequencies for each population, (2) specify a vector of sample sizes for each population, and (3) specify a string for filename for output. This step is baseline for all parameter sensitivity analyses.

```{r calc_af}
calc.allFreqs<-function(vcf,pop.list, pop.labs){
  allFreqs<-do.call(rbind,lapply(pop.list, function(pop){
    this.vcf<-cbind(vcf[,1:9],vcf$SNP,vcf[,grep(pop,colnames(vcf))])
    afs<-do.call(rbind,apply(this.vcf,1,calc.afs.vcf))
    return(afs$RefFreq)
  }))
  colnames(allFreqs)<-vcf$SNP
  rownames(allFreqs)<-pop.labs
  return(allFreqs)
}
#calculate selected allele frequencies
LG8<-calc.allFreqs(vcf[vcf$`#CHROM` == "LG8",],pop.list,pop.labs)
saveRDS(LG8,"dmc/selectedRegionAlleleFreqs_p4LG8.RDS")
saveRDS(vcf[vcf$`#CHROM`=="LG8","POS"],"dmc/selectedRegionPositions_p4LG8.RDS")

#calculate neutral allele frequencies
allFreqs<-calc.allFreqs(vcf[vcf$`#CHROM` %in% c("LG6","LG9","LG10","LG11","LG13","LG14","LG16","LG18","LG20"),],
                   pop.list,pop.labs)

```

Once those are calculated, I can calculate the neutral F matrix, in addition to its determinant and inverse.

```{r calc_neutralF}
sampleSizes<-unlist(lapply(pop.list,function(pop){
  n<-length(grep(pop,colnames(vcf)))
  return(2*n) }))
saveRDS(sampleSizes,"dmc/sampleSizes.RDS")
neutralF_filename<-"dmc/neutralF_p4LG8"
source("../programs/dmc-master/calcNeutralF.R")
F_estimate<-readRDS("dmc/neutralF_p4LG8.RDS")
det_FOmegas_neutral = det(Tmatrix %*% (F_estimate + sampleErrorMatrix) %*% t(Tmatrix))
saveRDS(det_FOmegas_neutral, "dmc/det_FOmegas_neutral_p4LG8.RDS")

inv_FOmegas_neutral = ginv(Tmatrix %*% (F_estimate + sampleErrorMatrix) %*% t(Tmatrix))
saveRDS(inv_FOmegas_neutral, "dmc/inv_FOmegas_neutral_p4LG8.RDS")

```

**Note: I had to use `r Sys.chmod('./dmc/')` to allow read/write in the scripts.**

Let's look at the allele frequencies on LG8. This will help inform the hypotheses I test with dmc.


```{r sig_alleleFreqs}
stacks.sig<-read.delim("p4.stacks.sig.snps.txt")
stacks.sig$SNP<-paste(stacks.sig$Chr,(stacks.sig$BP+1),sep=".")
# LG8[rownames(LG8) %in% c("FLFW","ALFW","LAFW","TXFW"),colnames(LG8) %in% stacks.sig$SNP]
# LG8[rownames(LG8) %in% c("FLFW","ALFW","LAFW","TXFW"),
#          which(colnames(LG8) %in% stacks.sig$SNP)[1]:
#            which(colnames(LG8) %in% stacks.sig$SNP)[length(which(colnames(LG8) %in% stacks.sig$SNP))]]
grp.colors<-c('#762a83','#af8dc3','#e7d4e8','#d9f0d3','#7fbf7b','#1b7837')
plot(LG8["TXFW",],col=grp.colors[1],pch="-",bty="L",lwd=2,cex=2)
points(LG8["LAFW",],col=grp.colors[2],pch="-",bty="L",lwd=2,cex=2)
points(LG8["ALFW",],col=grp.colors[3],pch="-",bty="L",lwd=2,cex=2)
points(LG8["FLFW",],col=grp.colors[6],pch="-",bty="L",lwd=2,cex=2)

cols<-colorRampPalette(t(brewer.pal(9,name="Blues")[3:7]))(100)

fields::image.plot(t(allFreqs[rownames(allFreqs) %in% c("FLFW","ALFW","LAFW","TXFW"),]),col=cols,
      axes=FALSE)
text(x=-0.02,y=1,"FLFW",srt=90,xpd=TRUE)
text(x=-0.02,y=0.67,"ALFW",srt=90,xpd=TRUE)
text(x=-0.02,y=0.35,"LAFW",srt=90,xpd=TRUE)
text(x=-0.02,y=0,"TXFW",srt=90,xpd=TRUE)
mtext("Position on LG8",1)
axis(3,at =t(LG8[rownames(LG8) %in% c("FLFW","ALFW","LAFW","TXFW"),colnames(LG8) %in% stacks.sig$SNP]),
     labels = FALSE,lwd=0,lwd.ticks = 1.75,tck=-.05)
```

This shows that the FLFW site has somewhat different allele frequencies than the other sites. This suggests that I should test to see whether there was an independent mutation in the FLFW population and the other three have a single sweep.

## Calculate F<sup>(S)</sup> matrices

We generate matrices for the following five models:
*1. All selected pops have independent mutations of beneficial allele
*2. All selected pops share beneficial allele via migration
*3. Beneficial allele was standing the ancestor of all selected pops
*4. There was an independent mutation in FLFW and the other three share a sweep.
*5. FLFW has an independent mutation and the others share a beneficial allele via standing variation

I've written a giant function to run all five models and calculate their composite likelihoods. The composite likelihoods include knowing how far away we are from the proposed selected site. Therefore, the first step for calculating them is to randomize with respect to the reference allele.  

The inverses and determinants generated can be used to calculate the log-likelihood of a site a given distance away from a proposed selected site. To do this, we sum the log-likelihoods across all SNPs in the window to obtain a composite log-likelihood under a given convergence model with a set of parameters for a proposed selected site. 

```{r runDMC,echo=TRUE}
```

I have to set some parameters before running the function, including the sets of parameters I want to try different settings for.

```{r setParameters}
```

And I'm going to run it with a variety of effective population sizes to see how robust the models are to that parameter.

```{r dmcNe, eval=FALSE}
```

I'm running this one on abies using `nohup Rscript ~/Projects/popgen/scripts/run_dmc.R > ~/Projects/popgen/dmc_ne.log 2>&1 &`. Process 1539 on 7 March 2018


##Plot maximum composite likelihood ratios for models over proposed selected sites

```{r plot_likelihoods}

#read in composite likelihood files and calculate max for all proposed selected sites
compLikelihood_neutral = readRDS("dmc/compLikelihood_neutral_p4LG8.RDS")
compLikelihood_neutral_site = sapply(1 : length(selSite), function(i) {
  max(unlist(compLikelihood_neutral[[i]]))
})

compLikelihood_ind = readRDS("dmc/compLikelihood_ind_p4LG8.RDS")
compLikelihood_ind_site = sapply(1 : length(selSite), function(i) {
  max(unlist(compLikelihood_ind[[i]]))
})

compLikelihood_mig = readRDS("dmc/compLikelihood_mig_p4LG8.RDS")
compLikelihood_mig_site = sapply(1 : length(selSite), function(i) {
  max(unlist(compLikelihood_mig[[i]]))
})

compLikelihood_sv = readRDS("dmc/compLikelihood_sv_p4LG8.RDS")
compLikelihood_sv_site = sapply(1 : length(selSite), function(i) {  
  max(unlist(compLikelihood_sv[[i]]))
})

compLikelihood_mixed_migInd = readRDS("dmc/compLikelihood_mixed_migInd_p4LG8.RDS")
compLikelihood_mixed_migInd_site = sapply(1 : length(selSite), function(i) {
  max(unlist(compLikelihood_mixed_migInd[[i]]))
})

compLikelihood_mixed_svInd = readRDS("dmc/compLikelihood_mixed_svInd_p4LG8.RDS")
compLikelihood_mixed_svInd_site = sapply(1 : length(selSite), function(i) {
  max(unlist(compLikelihood_mixed_svInd[[i]]))
})

plot_range = range(c((compLikelihood_ind_site - compLikelihood_neutral_site),
                     (compLikelihood_mig_site - compLikelihood_neutral_site), 
                     (compLikelihood_sv_site - compLikelihood_neutral_site), 
                     (compLikelihood_mixed_migInd_site - compLikelihood_neutral_site),
                     (compLikelihood_mixed_svInd_site - compLikelihood_neutral_site)))

plot(selSite, compLikelihood_ind_site - compLikelihood_neutral_site, type = "b",
     ylim = c(plot_range[1] - 50, plot_range[2] + 50),
     xlab = "Proposed position selected site",
     ylab = "Composite log-likelihood (model - neutral)")
lines(selSite, compLikelihood_mig_site - compLikelihood_neutral_site, col = "red",
      type = "b")
lines(selSite, compLikelihood_sv_site - compLikelihood_neutral_site, col = "blue",
      type = "b")
lines(selSite, compLikelihood_mixed_migInd_site - compLikelihood_neutral_site,
      col = "orange", lty = 2, type = "b")
lines(selSite, compLikelihood_mixed_svInd_site - compLikelihood_neutral_site,
      col = "green", lty = 2, type = "b")
legend("topright", col = c("black", "red", "blue", "orange", "green"),
       lty = c(rep(1, 3), rep(2, 2)), sapply(1 : 5, function(i) paste("Model", i)),
       cex = 0.5)


```

### Get maximum composite likelihood estimates

```{r mcle}
source("../programs/dmc-master/getMCLE.R")

#These won't work if the parameters aren't specified as above

## Model 1
getMCLEind(compLikelihood_ind, selSite, sels)

## Model 2
getMCLEmig(compLikelihood_mig, selSite, sels, migs, sources)

## Model 3
getMCLEsv_source(compLikelihood_sv, selSite, sels, gs, times, sources)

## Model 4
getMCLEmixed(compLikelihood_mixed_migInd, selSite, sels, gs[1], times[1], migs, sources)

## Model 5
getMCLEmixed(compLikelihood_mixed_svInd, selSite, sels, gs, times, migs[1], sources)

```

### Plot profile likelihood surfaces

Plot the profile likelihood surfaces for the parameter of time the beneficial allele has been standing independently in the selected populations for the models with the highest composite log-likelihoods.

<!-- ## Finding prolactin -->

<!-- ```{r pinpoint_prl,echo=FALSE,eval=FALSE} -->
<!-- #generate position data near prolactin -->

<!-- put.reg<-read.delim("putative.gene.regions.tsv",header=T) -->
<!-- prl<-put.reg[put.reg$Gene=="PRL" & put.reg$Chrom=="LG8",] -->
<!-- prl.loc<-as.numeric(prl$StartBP[2])+((as.numeric(prl$StopBP[2])-as.numeric(prl$StartBP[2]))/2) -->
<!-- ``` -->

